// @trigger.dev/core/dist/esm/v3/runEngineWorker/supervisor/http.js
import { z } from "zod";
import { WorkerApiConnectResponseBody, WorkerApiContinueRunExecutionRequestBody, WorkerApiDequeueResponseBody, WorkerApiHeartbeatResponseBody, WorkerApiRunAttemptCompleteResponseBody, WorkerApiRunAttemptStartResponseBody, WorkerApiRunHeartbeatResponseBody, WorkerApiRunLatestSnapshotResponseBody, WorkerApiSuspendRunResponseBody, WorkerApiRunSnapshotsSinceResponseBody, } from "./schemas.js";
import { getDefaultWorkerHeaders } from "./util.js";
import { wrapZodFetch } from "../../zodfetch.js";
import { createHeaders } from "../util.js";
import { WORKER_HEADERS } from "../consts.js";
import { SimpleStructuredLogger } from "../../utils/structuredLogger.js";
export class SupervisorHttpClient {
    apiUrl;
    workerToken;
    instanceName;
    defaultHeaders;
    sendRunDebugLogs;
    logger = new SimpleStructuredLogger("supervisor-http-client");
    constructor(opts) {
        this.apiUrl = opts.apiUrl.replace(/\/$/, "");
        this.workerToken = opts.workerToken;
        this.instanceName = opts.instanceName;
        this.defaultHeaders = getDefaultWorkerHeaders(opts);
        this.sendRunDebugLogs = opts.sendRunDebugLogs ?? false;
        if (!this.apiUrl) {
            throw new Error("apiURL is required and needs to be a non-empty string");
        }
        if (!this.workerToken) {
            throw new Error("workerToken is required and needs to be a non-empty string");
        }
        if (!this.instanceName) {
            throw new Error("instanceName is required and needs to be a non-empty string");
        }
    }
    async connect(body) {
        return wrapZodFetch(WorkerApiConnectResponseBody, `${this.apiUrl}/engine/v1/worker-actions/connect`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    async dequeue(body) {
        return wrapZodFetch(WorkerApiDequeueResponseBody, `${this.apiUrl}/engine/v1/worker-actions/dequeue`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    /** @deprecated Not currently used */
    async dequeueFromVersion(deploymentId, maxRunCount = 1, runnerId) {
        return wrapZodFetch(WorkerApiDequeueResponseBody, `${this.apiUrl}/engine/v1/worker-actions/deployments/${deploymentId}/dequeue?maxRunCount=${maxRunCount}`, {
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
        });
    }
    async heartbeatWorker(body) {
        return wrapZodFetch(WorkerApiHeartbeatResponseBody, `${this.apiUrl}/engine/v1/worker-actions/heartbeat`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    async heartbeatRun(runId, snapshotId, body, runnerId) {
        return wrapZodFetch(WorkerApiRunHeartbeatResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/${snapshotId}/heartbeat`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    async startRunAttempt(runId, snapshotId, body, runnerId) {
        return wrapZodFetch(WorkerApiRunAttemptStartResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/${snapshotId}/attempts/start`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
            body: JSON.stringify(body),
        });
    }
    async completeRunAttempt(runId, snapshotId, body, runnerId) {
        return wrapZodFetch(WorkerApiRunAttemptCompleteResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/${snapshotId}/attempts/complete`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
            body: JSON.stringify(body),
        });
    }
    async getLatestSnapshot(runId, runnerId) {
        return wrapZodFetch(WorkerApiRunLatestSnapshotResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/latest`, {
            method: "GET",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
        });
    }
    async getSnapshotsSince(runId, snapshotId, runnerId) {
        return wrapZodFetch(WorkerApiRunSnapshotsSinceResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/since/${snapshotId}`, {
            method: "GET",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
        });
    }
    async sendDebugLog(runId, body, runnerId) {
        if (!this.sendRunDebugLogs) {
            return;
        }
        try {
            const res = await wrapZodFetch(z.unknown(), `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/logs/debug`, {
                method: "POST",
                headers: {
                    ...this.defaultHeaders,
                    ...this.runnerIdHeader(runnerId),
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(body),
            });
            if (!res.success) {
                this.logger.error("Failed to send debug log", { error: res.error });
            }
        }
        catch (error) {
            this.logger.error("Failed to send debug log (caught error)", { error });
        }
    }
    async continueRunExecution(runId, snapshotId, runnerId) {
        return wrapZodFetch(WorkerApiContinueRunExecutionRequestBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/${snapshotId}/continue`, {
            method: "GET",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
            },
        });
    }
    async submitSuspendCompletion({ runId, snapshotId, runnerId, body, }) {
        return wrapZodFetch(WorkerApiSuspendRunResponseBody, `${this.apiUrl}/engine/v1/worker-actions/runs/${runId}/snapshots/${snapshotId}/suspend`, {
            method: "POST",
            headers: {
                ...this.defaultHeaders,
                ...this.runnerIdHeader(runnerId),
                "Content-Type": "application/json",
            },
            body: JSON.stringify(body),
        });
    }
    runnerIdHeader(runnerId) {
        return createHeaders({
            [WORKER_HEADERS.RUNNER_ID]: runnerId,
        });
    }
}
