// @trigger.dev/core/dist/esm/v3/inputStreams/index.js
import { getGlobal, registerGlobal } from "../utils/globals.js";
import { NoopInputStreamManager } from "./noopManager.js";
const API_NAME = "input-streams";
const NOOP_MANAGER = new NoopInputStreamManager();
export class InputStreamsAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new InputStreamsAPI();
        }
        return this._instance;
    }
    setGlobalManager(manager) {
        return registerGlobal(API_NAME, manager);
    }
    #getManager() {
        return getGlobal(API_NAME) ?? NOOP_MANAGER;
    }
    setRunId(runId, streamsVersion) {
        this.#getManager().setRunId(runId, streamsVersion);
    }
    on(streamId, handler) {
        return this.#getManager().on(streamId, handler);
    }
    once(streamId, options) {
        return this.#getManager().once(streamId, options);
    }
    peek(streamId) {
        return this.#getManager().peek(streamId);
    }
    lastSeqNum(streamId) {
        return this.#getManager().lastSeqNum(streamId);
    }
    clearHandlers() {
        this.#getManager().clearHandlers();
    }
    reset() {
        this.#getManager().reset();
    }
    disconnect() {
        this.#getManager().disconnect();
    }
    connectTail(runId, fromSeq) {
        this.#getManager().connectTail(runId, fromSeq);
    }
}
