// @trigger.dev/core/dist/esm/v3/schemas/resources.js
import { z } from "zod";
import { QueueManifest, RetryOptions, ScheduleMetadata } from "./schemas.js";
import { MachineConfig } from "./common.js";
export const TaskResource = z.object({
    id: z.string(),
    description: z.string().optional(),
    filePath: z.string(),
    exportName: z.string().optional(),
    queue: QueueManifest.extend({ name: z.string().optional() }).optional(),
    retry: RetryOptions.optional(),
    machine: MachineConfig.optional(),
    triggerSource: z.string().optional(),
    schedule: ScheduleMetadata.optional(),
    maxDuration: z.number().optional(),
    ttl: z.string().or(z.number().nonnegative().int()).optional(),
    // JSONSchema type - using z.unknown() for runtime validation to accept JSONSchema7
    payloadSchema: z.unknown().optional(),
});
export const BackgroundWorkerSourceFileMetadata = z.object({
    filePath: z.string(),
    contents: z.string(),
    contentHash: z.string(),
    taskIds: z.array(z.string()),
});
export const PromptResource = z.object({
    id: z.string(),
    description: z.string().optional(),
    filePath: z.string(),
    exportName: z.string().optional(),
    /** The default template content */
    content: z.string().optional(),
    /** Default model identifier */
    model: z.string().optional(),
    /** Default model config (temperature, maxTokens, etc.) */
    config: z.record(z.unknown()).optional(),
    /** JSONSchema7 for template variables */
    variableSchema: z.unknown().optional(),
});
export const BackgroundWorkerMetadata = z.object({
    packageVersion: z.string(),
    contentHash: z.string(),
    cliPackageVersion: z.string().optional(),
    tasks: z.array(TaskResource),
    prompts: z.array(PromptResource).optional(),
    queues: z.array(QueueManifest).optional(),
    sourceFiles: z.array(BackgroundWorkerSourceFileMetadata).optional(),
    runtime: z.string().optional(),
    runtimeVersion: z.string().optional(),
});
export const ImageDetailsMetadata = z.object({
    contentHash: z.string(),
    imageTag: z.string(),
});
