// @trigger.dev/core/dist/commonjs/v3/runEngineWorker/workload/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkloadRunSnapshotsSinceResponseBody = exports.WorkloadDequeueFromVersionResponseBody = exports.WorkloadDebugLogRequestBody = exports.WorkloadRunLatestSnapshotResponseBody = exports.WorkloadRunAttemptStartResponseBody = exports.WorkloadRunAttemptStartRequestBody = exports.WorkloadRunAttemptCompleteResponseBody = exports.WorkloadRunAttemptCompleteRequestBody = exports.WorkloadContinueRunExecutionResponseBody = exports.WorkloadSuspendRunResponseBody = exports.WorkloadHeartbeatResponseBody = exports.WorkloadHeartbeatRequestBody = void 0;
const zod_1 = require("zod");
const schemas_js_1 = require("../supervisor/schemas.js");
exports.WorkloadHeartbeatRequestBody = schemas_js_1.WorkerApiRunHeartbeatRequestBody;
exports.WorkloadHeartbeatResponseBody = schemas_js_1.WorkerApiHeartbeatResponseBody;
exports.WorkloadSuspendRunResponseBody = zod_1.z.discriminatedUnion("ok", [
    zod_1.z.object({
        ok: zod_1.z.literal(true),
    }),
    zod_1.z.object({
        ok: zod_1.z.literal(false),
        error: zod_1.z.string(),
    }),
]);
exports.WorkloadContinueRunExecutionResponseBody = schemas_js_1.WorkerApiContinueRunExecutionRequestBody;
exports.WorkloadRunAttemptCompleteRequestBody = schemas_js_1.WorkerApiRunAttemptCompleteRequestBody;
exports.WorkloadRunAttemptCompleteResponseBody = schemas_js_1.WorkerApiRunAttemptCompleteResponseBody;
exports.WorkloadRunAttemptStartRequestBody = schemas_js_1.WorkerApiRunAttemptStartRequestBody;
exports.WorkloadRunAttemptStartResponseBody = schemas_js_1.WorkerApiRunAttemptStartResponseBody;
exports.WorkloadRunLatestSnapshotResponseBody = schemas_js_1.WorkerApiRunLatestSnapshotResponseBody;
exports.WorkloadDebugLogRequestBody = schemas_js_1.WorkerApiDebugLogBody;
exports.WorkloadDequeueFromVersionResponseBody = schemas_js_1.WorkerApiDequeueFromVersionResponseBody;
exports.WorkloadRunSnapshotsSinceResponseBody = schemas_js_1.WorkerApiRunSnapshotsSinceResponseBody;
