// @trigger.dev/core/dist/commonjs/v3/resource-catalog/noopResourceCatalog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopResourceCatalog = void 0;
class NoopResourceCatalog {
    registerTaskMetadata(task) {
        // noop
    }
    setCurrentFileContext(filePath, entryPoint) {
        // noop
    }
    clearCurrentFileContext() {
        // noop
    }
    updateTaskMetadata(id, updates) {
        // noop
    }
    listTaskManifests() {
        return [];
    }
    getTaskManifest(id) {
        return undefined;
    }
    getTask(id) {
        return undefined;
    }
    getTaskSchema(id) {
        return undefined;
    }
    taskExists(id) {
        return false;
    }
    disable() {
        // noop
    }
    registerWorkerManifest(workerManifest) {
        // noop
    }
    registerQueueMetadata(queue) {
        // noop
    }
    listQueueManifests() {
        return [];
    }
    registerPromptMetadata(prompt) {
        // noop
    }
    listPromptManifests() {
        return [];
    }
    getPrompt(id) {
        return undefined;
    }
    getPromptSchema(id) {
        return undefined;
    }
}
exports.NoopResourceCatalog = NoopResourceCatalog;
