// @trigger.dev/core/dist/commonjs/v3/inputStreams/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputStreamsAPI = void 0;
const globals_js_1 = require("../utils/globals.js");
const noopManager_js_1 = require("./noopManager.js");
const API_NAME = "input-streams";
const NOOP_MANAGER = new noopManager_js_1.NoopInputStreamManager();
class InputStreamsAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new InputStreamsAPI();
        }
        return this._instance;
    }
    setGlobalManager(manager) {
        return (0, globals_js_1.registerGlobal)(API_NAME, manager);
    }
    #getManager() {
        return (0, globals_js_1.getGlobal)(API_NAME) ?? NOOP_MANAGER;
    }
    setRunId(runId, streamsVersion) {
        this.#getManager().setRunId(runId, streamsVersion);
    }
    on(streamId, handler) {
        return this.#getManager().on(streamId, handler);
    }
    once(streamId, options) {
        return this.#getManager().once(streamId, options);
    }
    peek(streamId) {
        return this.#getManager().peek(streamId);
    }
    lastSeqNum(streamId) {
        return this.#getManager().lastSeqNum(streamId);
    }
    clearHandlers() {
        this.#getManager().clearHandlers();
    }
    reset() {
        this.#getManager().reset();
    }
    disconnect() {
        this.#getManager().disconnect();
    }
    connectTail(runId, fromSeq) {
        this.#getManager().connectTail(runId, fromSeq);
    }
}
exports.InputStreamsAPI = InputStreamsAPI;
