// @trigger.dev/core/dist/commonjs/v3/schemas/api-type.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const api_js_1 = require("./api.js");
(0, vitest_1.describe)("InitializeDeploymentRequestBody", () => {
    const base = { contentHash: "abc123" };
    (0, vitest_1.describe)("non-native build variant (isNativeBuild omitted or false)", () => {
        (0, vitest_1.it)("parses with only required fields", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse(base);
            (0, vitest_1.expect)(result.success).toBe(true);
            if (result.success) {
                (0, vitest_1.expect)(result.data.isNativeBuild).toBe(false);
            }
        });
        (0, vitest_1.it)("parses with isNativeBuild explicitly false", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({ ...base, isNativeBuild: false });
            (0, vitest_1.expect)(result.success).toBe(true);
        });
        (0, vitest_1.it)("parses with optional base fields", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                userId: "user_1",
                type: "MANAGED",
                runtime: "node",
                initialStatus: "PENDING",
            });
            (0, vitest_1.expect)(result.success).toBe(true);
        });
        (0, vitest_1.it)("strips native-only fields when isNativeBuild is false", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                isNativeBuild: false,
                skipPromotion: true,
            });
            // Zod discriminatedUnion matches the non-native branch and strips unknown keys
            (0, vitest_1.expect)(result.success).toBe(true);
            if (result.success) {
                (0, vitest_1.expect)("skipPromotion" in result.data).toBe(false);
            }
        });
    });
    (0, vitest_1.describe)("native build variant (isNativeBuild: true)", () => {
        (0, vitest_1.it)("parses with isNativeBuild true", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                isNativeBuild: true,
            });
            (0, vitest_1.expect)(result.success).toBe(true);
            if (result.success) {
                (0, vitest_1.expect)(result.data.isNativeBuild).toBe(true);
            }
        });
        (0, vitest_1.it)("parses with native-specific optional fields", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                isNativeBuild: true,
                skipPromotion: true,
                artifactKey: "artifact_abc",
                configFilePath: "trigger.config.ts",
                skipEnqueue: true,
            });
            (0, vitest_1.expect)(result.success).toBe(true);
            if (result.success) {
                (0, vitest_1.expect)(result.data.skipPromotion).toBe(true);
                (0, vitest_1.expect)(result.data.artifactKey).toBe("artifact_abc");
                (0, vitest_1.expect)(result.data.configFilePath).toBe("trigger.config.ts");
                (0, vitest_1.expect)(result.data.skipEnqueue).toBe(true);
            }
        });
        (0, vitest_1.it)("skipEnqueue defaults to false when omitted", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                isNativeBuild: true,
            });
            (0, vitest_1.expect)(result.success).toBe(true);
            if (result.success) {
                (0, vitest_1.expect)(result.data.skipEnqueue).toBe(false);
            }
        });
    });
    (0, vitest_1.describe)("rejects invalid inputs", () => {
        (0, vitest_1.it)("rejects missing contentHash", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({});
            (0, vitest_1.expect)(result.success).toBe(false);
        });
        (0, vitest_1.it)("rejects invalid type enum value", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                type: "INVALID",
            });
            (0, vitest_1.expect)(result.success).toBe(false);
        });
        (0, vitest_1.it)("rejects invalid initialStatus enum value", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.safeParse({
                ...base,
                initialStatus: "RUNNING",
            });
            (0, vitest_1.expect)(result.success).toBe(false);
        });
    });
    (0, vitest_1.describe)("type-level checks", () => {
        (0, vitest_1.it)("native variant exposes native-specific fields", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.parse({
                ...base,
                isNativeBuild: true,
                skipPromotion: true,
            });
            if (result.isNativeBuild === true) {
                const _skipPromotion = result.skipPromotion;
                const _artifactKey = result.artifactKey;
                const _configFilePath = result.configFilePath;
                (0, vitest_1.expect)(_skipPromotion).toBe(true);
                (0, vitest_1.expect)(_artifactKey).toBeUndefined();
                (0, vitest_1.expect)(_configFilePath).toBeUndefined();
            }
        });
        (0, vitest_1.it)("non-native variant narrows correctly", () => {
            const result = api_js_1.InitializeDeploymentRequestBody.parse(base);
            if (!result.isNativeBuild) {
                // Should only have base fields — native-specific fields should not exist
                const narrowed = result;
                (0, vitest_1.expect)(narrowed.contentHash).toBe("abc123");
            }
        });
    });
});
