// @trigger.dev/core/dist/commonjs/v3/schemas/checkpoints.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckpointServiceRestoreRequestBody = exports.CheckpointServiceSuspendResponseBody = exports.CheckpointServiceSuspendRequestBody = void 0;
const runEngine_js_1 = require("./runEngine.js");
const zod_1 = __importDefault(require("zod"));
const CallbackUrl = zod_1.default
    .string()
    .url()
    .transform((url) => new URL(url));
exports.CheckpointServiceSuspendRequestBody = zod_1.default.object({
    type: runEngine_js_1.CheckpointType,
    runId: zod_1.default.string(),
    snapshotId: zod_1.default.string(),
    runnerId: zod_1.default.string(),
    projectRef: zod_1.default.string(),
    deploymentVersion: zod_1.default.string(),
    reason: zod_1.default.string().optional(),
});
exports.CheckpointServiceSuspendResponseBody = zod_1.default.object({
    ok: zod_1.default.literal(true),
});
exports.CheckpointServiceRestoreRequestBody = runEngine_js_1.DequeuedMessage.required({ checkpoint: true });
