// @trigger.dev/core/dist/commonjs/v3/schemas/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runFriendlyStatus = exports.ListDashboardsResponseBody = exports.DashboardSummary = exports.DashboardWidgetSummary = exports.QuerySchemaResponseBody = exports.QuerySchemaTable = exports.QuerySchemaColumn = exports.QueryExecuteResponseBody = exports.QueryExecuteCSVResponseBody = exports.QueryExecuteJSONResponseBody = exports.QueryExecuteRequestBody = void 0;
const zod_1 = require("zod");
/**
 * Request body schema for executing a query
 */
exports.QueryExecuteRequestBody = zod_1.z.object({
    query: zod_1.z.string(),
    scope: zod_1.z.enum(["organization", "project", "environment"]).default("environment"),
    period: zod_1.z.string().nullish(),
    from: zod_1.z.string().nullish(),
    to: zod_1.z.string().nullish(),
    format: zod_1.z.enum(["json", "csv"]).default("json"),
});
/**
 * Response body schema for JSON format queries
 */
exports.QueryExecuteJSONResponseBody = zod_1.z.object({
    format: zod_1.z.literal("json"),
    results: zod_1.z.array(zod_1.z.record(zod_1.z.any())),
});
/**
 * Response body type for CSV format queries
 */
exports.QueryExecuteCSVResponseBody = zod_1.z.object({
    format: zod_1.z.literal("csv"),
    results: zod_1.z.string(),
});
exports.QueryExecuteResponseBody = zod_1.z.discriminatedUnion("format", [
    exports.QueryExecuteJSONResponseBody,
    exports.QueryExecuteCSVResponseBody,
]);
// ---------------------------------------------------------------------------
// Query schema discovery types
// ---------------------------------------------------------------------------
/**
 * Schema for a single column in the query schema response
 */
exports.QuerySchemaColumn = zod_1.z.object({
    name: zod_1.z.string(),
    type: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    example: zod_1.z.string().optional(),
    allowedValues: zod_1.z.array(zod_1.z.string()).optional(),
    coreColumn: zod_1.z.boolean().optional(),
});
/**
 * Schema for a single table in the query schema response
 */
exports.QuerySchemaTable = zod_1.z.object({
    name: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    timeColumn: zod_1.z.string().optional(),
    columns: zod_1.z.array(exports.QuerySchemaColumn),
});
/**
 * Response body for the query schema endpoint
 */
exports.QuerySchemaResponseBody = zod_1.z.object({
    tables: zod_1.z.array(exports.QuerySchemaTable),
});
// ---------------------------------------------------------------------------
// Dashboard types
// ---------------------------------------------------------------------------
/**
 * Summary of a widget in a dashboard (for listing)
 */
exports.DashboardWidgetSummary = zod_1.z.object({
    id: zod_1.z.string(),
    title: zod_1.z.string(),
    query: zod_1.z.string(),
    type: zod_1.z.enum(["bignumber", "chart", "table", "title"]),
});
/**
 * Summary of a dashboard (for listing)
 */
exports.DashboardSummary = zod_1.z.object({
    key: zod_1.z.string(),
    title: zod_1.z.string(),
    widgets: zod_1.z.array(exports.DashboardWidgetSummary),
});
/**
 * Response body for listing dashboards
 */
exports.ListDashboardsResponseBody = zod_1.z.object({
    dashboards: zod_1.z.array(exports.DashboardSummary),
});
// ---------------------------------------------------------------------------
// Query table row types
// ---------------------------------------------------------------------------
/**
 * User-facing friendly run status values returned by the query system.
 */
exports.runFriendlyStatus = [
    "Delayed",
    "Queued",
    "Pending version",
    "Dequeued",
    "Executing",
    "Waiting",
    "Reattempting",
    "Paused",
    "Canceled",
    "Interrupted",
    "Completed",
    "Failed",
    "System failure",
    "Crashed",
    "Expired",
    "Timed out",
];
