// @trigger.dev/core/dist/commonjs/v3/schemas/resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageDetailsMetadata = exports.BackgroundWorkerMetadata = exports.PromptResource = exports.BackgroundWorkerSourceFileMetadata = exports.TaskResource = void 0;
const zod_1 = require("zod");
const schemas_js_1 = require("./schemas.js");
const common_js_1 = require("./common.js");
exports.TaskResource = zod_1.z.object({
    id: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    filePath: zod_1.z.string(),
    exportName: zod_1.z.string().optional(),
    queue: schemas_js_1.QueueManifest.extend({ name: zod_1.z.string().optional() }).optional(),
    retry: schemas_js_1.RetryOptions.optional(),
    machine: common_js_1.MachineConfig.optional(),
    triggerSource: zod_1.z.string().optional(),
    schedule: schemas_js_1.ScheduleMetadata.optional(),
    maxDuration: zod_1.z.number().optional(),
    ttl: zod_1.z.string().or(zod_1.z.number().nonnegative().int()).optional(),
    // JSONSchema type - using z.unknown() for runtime validation to accept JSONSchema7
    payloadSchema: zod_1.z.unknown().optional(),
});
exports.BackgroundWorkerSourceFileMetadata = zod_1.z.object({
    filePath: zod_1.z.string(),
    contents: zod_1.z.string(),
    contentHash: zod_1.z.string(),
    taskIds: zod_1.z.array(zod_1.z.string()),
});
exports.PromptResource = zod_1.z.object({
    id: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    filePath: zod_1.z.string(),
    exportName: zod_1.z.string().optional(),
    /** The default template content */
    content: zod_1.z.string().optional(),
    /** Default model identifier */
    model: zod_1.z.string().optional(),
    /** Default model config (temperature, maxTokens, etc.) */
    config: zod_1.z.record(zod_1.z.unknown()).optional(),
    /** JSONSchema7 for template variables */
    variableSchema: zod_1.z.unknown().optional(),
});
exports.BackgroundWorkerMetadata = zod_1.z.object({
    packageVersion: zod_1.z.string(),
    contentHash: zod_1.z.string(),
    cliPackageVersion: zod_1.z.string().optional(),
    tasks: zod_1.z.array(exports.TaskResource),
    prompts: zod_1.z.array(exports.PromptResource).optional(),
    queues: zod_1.z.array(schemas_js_1.QueueManifest).optional(),
    sourceFiles: zod_1.z.array(exports.BackgroundWorkerSourceFileMetadata).optional(),
    runtime: zod_1.z.string().optional(),
    runtimeVersion: zod_1.z.string().optional(),
});
exports.ImageDetailsMetadata = zod_1.z.object({
    contentHash: zod_1.z.string(),
    imageTag: zod_1.z.string(),
});
