// @trigger.dev/core/dist/commonjs/v3/schemas/webhooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = exports.AlertWebhookErrorGroupObject = exports.AlertWebhookDeploymentFailedObject = exports.AlertWebhookDeploymentSuccessObject = exports.DeployError = void 0;
const zod_1 = require("zod");
const api_js_1 = require("./api.js");
const common_js_1 = require("./common.js");
/** Represents a failed run alert webhook payload */
const AlertWebhookRunFailedObject = zod_1.z.object({
    /** Task information */
    task: zod_1.z.object({
        /** Unique identifier for the task */
        id: zod_1.z.string(),
        /** File path where the task is defined */
        filePath: zod_1.z.string(),
        /** Name of the exported task function */
        exportName: zod_1.z.string().optional(),
        /** Version of the task */
        version: zod_1.z.string(),
        /** Version of the SDK used */
        sdkVersion: zod_1.z.string(),
        /** Version of the CLI used */
        cliVersion: zod_1.z.string(),
    }),
    /** Run information */
    run: zod_1.z.object({
        /** Unique identifier for the run */
        id: zod_1.z.string(),
        /** Run number */
        number: zod_1.z.number(),
        /** Current status of the run */
        status: api_js_1.RunStatus,
        /** When the run was created */
        createdAt: zod_1.z.coerce.date(),
        /** When the run started executing */
        startedAt: zod_1.z.coerce.date().optional(),
        /** When the run finished executing */
        completedAt: zod_1.z.coerce.date().optional(),
        /** Whether this is a test run */
        isTest: zod_1.z.boolean(),
        /** Idempotency key for the run */
        idempotencyKey: zod_1.z.string().optional(),
        /** Associated tags */
        tags: zod_1.z.array(zod_1.z.string()),
        /** Error information */
        error: common_js_1.TaskRunError,
        /** Whether the run was an out-of-memory error */
        isOutOfMemoryError: zod_1.z.boolean(),
        /** Machine preset used for the run */
        machine: zod_1.z.string(),
        /** URL to view the run in the dashboard */
        dashboardUrl: zod_1.z.string(),
    }),
    /** Environment information */
    environment: zod_1.z.object({
        /** Environment ID */
        id: zod_1.z.string(),
        /** Environment type */
        type: common_js_1.RuntimeEnvironmentTypeSchema,
        /** Environment slug */
        slug: zod_1.z.string(),
        /** Environment branch name */
        branchName: zod_1.z.string().optional(),
    }),
    /** Organization information */
    organization: zod_1.z.object({
        /** Organization ID */
        id: zod_1.z.string(),
        /** Organization slug */
        slug: zod_1.z.string(),
        /** Organization name */
        name: zod_1.z.string(),
    }),
    /** Project information */
    project: zod_1.z.object({
        /** Project ID */
        id: zod_1.z.string(),
        /** Project reference */
        ref: zod_1.z.string(),
        /** Project slug */
        slug: zod_1.z.string(),
        /** Project name */
        name: zod_1.z.string(),
    }),
});
/** Represents a deployment error */
exports.DeployError = zod_1.z.object({
    /** Error name */
    name: zod_1.z.string(),
    /** Error message */
    message: zod_1.z.string(),
    /** Error stack trace */
    stack: zod_1.z.string().optional(),
    /** Standard error output */
    stderr: zod_1.z.string().optional(),
});
const deploymentCommonProperties = {
    /** Environment information */
    environment: zod_1.z.object({
        id: zod_1.z.string(),
        type: common_js_1.RuntimeEnvironmentTypeSchema,
        slug: zod_1.z.string(),
        /** Environment branch name */
        branchName: zod_1.z.string().optional(),
    }),
    /** Organization information */
    organization: zod_1.z.object({
        id: zod_1.z.string(),
        slug: zod_1.z.string(),
        name: zod_1.z.string(),
    }),
    /** Project information */
    project: zod_1.z.object({
        id: zod_1.z.string(),
        ref: zod_1.z.string(),
        slug: zod_1.z.string(),
        name: zod_1.z.string(),
    }),
    /** Git metadata for the deployment source code */
    git: zod_1.z
        .object({
        branch: zod_1.z.string(),
        commitSha: zod_1.z.string(),
        commitMessage: zod_1.z.string(),
        commitUrl: zod_1.z.string(),
        branchUrl: zod_1.z.string(),
        pullRequestNumber: zod_1.z.number().optional(),
        pullRequestTitle: zod_1.z.string().optional(),
        pullRequestUrl: zod_1.z.string().optional(),
        provider: zod_1.z.string().optional(),
    })
        .optional(),
    /** Vercel integration data */
    vercel: zod_1.z
        .object({
        deploymentUrl: zod_1.z.string(),
    })
        .optional(),
};
const deploymentDeploymentCommonProperties = {
    /** Deployment ID */
    id: zod_1.z.string(),
    /** Deployment status */
    status: zod_1.z.string(),
    /** Deployment version */
    version: zod_1.z.string(),
    /** Short code identifier */
    shortCode: zod_1.z.string(),
};
/** Represents a successful deployment alert webhook payload */
exports.AlertWebhookDeploymentSuccessObject = zod_1.z.object({
    ...deploymentCommonProperties,
    deployment: zod_1.z.object({
        ...deploymentDeploymentCommonProperties,
        /** When the deployment completed */
        deployedAt: zod_1.z.coerce.date(),
    }),
    /** Deployed tasks */
    tasks: zod_1.z.array(zod_1.z.object({
        /** Task ID */
        id: zod_1.z.string(),
        /** File path where the task is defined */
        filePath: zod_1.z.string(),
        /** Name of the exported task function */
        exportName: zod_1.z.string().optional(),
        /** Source of the trigger */
        triggerSource: zod_1.z.string(),
    })),
});
/** Represents a failed deployment alert webhook payload */
exports.AlertWebhookDeploymentFailedObject = zod_1.z.object({
    ...deploymentCommonProperties,
    deployment: zod_1.z.object({
        ...deploymentDeploymentCommonProperties,
        /** When the deployment failed */
        failedAt: zod_1.z.coerce.date(),
    }),
    /** Error information */
    error: exports.DeployError,
});
/** Represents an error group alert webhook payload */
exports.AlertWebhookErrorGroupObject = zod_1.z.object({
    /** Classification of the error alert */
    classification: zod_1.z.enum(["new_issue", "regression", "unignored"]),
    /** Error information */
    error: zod_1.z.object({
        /** Error fingerprint identifier */
        fingerprint: zod_1.z.string(),
        /** Error type */
        type: zod_1.z.string(),
        /** Error message */
        message: zod_1.z.string(),
        /** Sample stack trace */
        stackTrace: zod_1.z.string().optional(),
        /** When the error was first seen */
        firstSeen: zod_1.z.coerce.date(),
        /** When the error was last seen */
        lastSeen: zod_1.z.coerce.date(),
        /** Number of occurrences */
        occurrenceCount: zod_1.z.number(),
        /** Task identifier where the error occurred */
        taskIdentifier: zod_1.z.string(),
    }),
    /** Environment information */
    environment: zod_1.z.object({
        /** Environment ID */
        id: zod_1.z.string(),
        /** Environment name */
        name: zod_1.z.string(),
    }),
    /** Organization information */
    organization: zod_1.z.object({
        /** Organization ID */
        id: zod_1.z.string(),
        /** Organization slug */
        slug: zod_1.z.string(),
        /** Organization name */
        name: zod_1.z.string(),
    }),
    /** Project information */
    project: zod_1.z.object({
        /** Project ID */
        id: zod_1.z.string(),
        /** Project reference */
        ref: zod_1.z.string(),
        /** Project slug */
        slug: zod_1.z.string(),
        /** Project name */
        name: zod_1.z.string(),
    }),
    /** URL to view the error in the dashboard */
    dashboardUrl: zod_1.z.string(),
});
/** Common properties for all webhooks */
const commonProperties = {
    /** Webhook ID */
    id: zod_1.z.string(),
    /** When the webhook was created */
    created: zod_1.z.coerce.date(),
    /** Version of the webhook */
    webhookVersion: zod_1.z.string(),
};
/** Represents all possible webhook types */
exports.Webhook = zod_1.z.discriminatedUnion("type", [
    /** Run failed alert webhook */
    zod_1.z.object({
        ...commonProperties,
        type: zod_1.z.literal("alert.run.failed"),
        object: AlertWebhookRunFailedObject,
    }),
    /** Deployment success alert webhook */
    zod_1.z.object({
        ...commonProperties,
        type: zod_1.z.literal("alert.deployment.success"),
        object: exports.AlertWebhookDeploymentSuccessObject,
    }),
    /** Deployment failed alert webhook */
    zod_1.z.object({
        ...commonProperties,
        type: zod_1.z.literal("alert.deployment.failed"),
        object: exports.AlertWebhookDeploymentFailedObject,
    }),
    /** Error group alert webhook */
    zod_1.z.object({
        ...commonProperties,
        type: zod_1.z.literal("alert.error"),
        object: exports.AlertWebhookErrorGroupObject,
    }),
]);
