// @s2-dev/streamstore/dist/cjs/lib/stream/transport/fetch/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchTransport = exports.FetchAppendSession = exports.FetchReadSession = void 0;
const debug_1 = require("debug");
const error_js_1 = require("../../../../error.js");
const index_js_1 = require("../../../../generated/client/index.js");
const index_js_2 = require("../../../../generated/index.js");
const utils_js_1 = require("../../../../utils.js");
const base64_js_1 = require("../../../base64.js");
const event_stream_js_1 = require("../../../event-stream.js");
const Redacted = require("../../../redacted.js");
const result_js_1 = require("../../../result.js");
const retry_js_1 = require("../../../retry.js");
const runtime_js_1 = require("../../runtime.js");
const shared_js_1 = require("./shared.js");
const debug = (0, debug_1.default)("s2:fetch");
class FetchReadSession extends ReadableStream {
    static async create(client, name, args, options) {
        debug("FetchReadSession.create stream=%s args=%o", name, args);
        const { as, ignore_command_records, ...queryParams } = args ?? {};
        const response = await (0, index_js_2.read)({
            client,
            path: {
                stream: name,
            },
            headers: {
                accept: "text/event-stream",
                ...(as === "bytes" ? { "s2-format": "base64" } : {}),
            },
            query: queryParams,
            parseAs: "stream",
            ...options,
        });
        if (response.error) {
            // Convert error to S2Error and throw - let caller handle connection errors
            const status = response.response.status;
            const error = "message" in response.error
                ? new error_js_1.S2Error({
                    message: response.error.message,
                    code: response.error.code ?? undefined,
                    status,
                })
                : status === 416
                    ? new error_js_1.RangeNotSatisfiableError({
                        status,
                    })
                    : new error_js_1.S2Error({
                        message: response.response.statusText ?? "Request failed",
                        status,
                    });
            throw error;
        }
        if (!response.response.body) {
            throw new error_js_1.S2Error({
                message: "No body in SSE response",
                code: "INVALID_RESPONSE",
                status: 502,
                origin: "sdk",
            });
        }
        const format = (args?.as ?? "string");
        return new FetchReadSession(response.response.body, format);
    }
    _nextReadPosition = undefined;
    _lastObservedTail = undefined;
    constructor(stream, format) {
        // Track error from parser
        let parserError = null;
        // Track last ping time for timeout detection (20s without a ping = timeout)
        let lastPingTimeMs = performance.now();
        const PING_TIMEOUT_MS = 20000; // 20 seconds
        // Create EventStream that parses SSE and yields records
        const eventStream = new event_stream_js_1.EventStream(stream, (msg) => {
            // Update ping time on ANY event from the server
            lastPingTimeMs = performance.now();
            // Parse SSE events according to the S2 protocol
            if (msg.event === "batch" && msg.data) {
                const rawBatch = JSON.parse(msg.data);
                const batch = (() => {
                    // If format is bytes, decode base64 to Uint8Array
                    if (format === "bytes") {
                        return {
                            ...rawBatch,
                            records: rawBatch.records.map((record) => ({
                                ...record,
                                body: record.body ? (0, base64_js_1.decodeFromBase64)(record.body) : undefined,
                                headers: record.headers?.map((header) => header.map((h) => (0, base64_js_1.decodeFromBase64)(h))),
                            })),
                        };
                    }
                    else {
                        return {
                            ...rawBatch,
                            records: rawBatch.records.map((record) => ({
                                ...record,
                                headers: record.headers || undefined,
                            })),
                        };
                    }
                })();
                if (batch.tail) {
                    this._lastObservedTail = batch.tail;
                }
                let lastRecord = batch.records?.at(-1);
                if (lastRecord) {
                    this._nextReadPosition = {
                        seq_num: lastRecord.seq_num + 1,
                        timestamp: lastRecord.timestamp,
                    };
                }
                return { done: false, batch: true, value: batch.records ?? [] };
            }
            if (msg.event === "error") {
                // Store error and signal end of stream
                // SSE error events are server errors - treat as 503 (Service Unavailable) for retry logic
                debug("parse event error");
                parserError = new error_js_1.S2Error({
                    message: msg.data ?? "Unknown error",
                    status: 503,
                });
                return { done: true };
            }
            if (msg.event === "ping") {
                debug("ping");
            }
            // Skip ping events and other events
            return { done: false };
        });
        // Wrap the EventStream to convert records to ReadResult and check for errors
        const reader = eventStream.getReader();
        let done = false;
        let pendingRead = null;
        super({
            pull: async (controller) => {
                // Loop to retry when stale timeouts fire
                while (true) {
                    if (done) {
                        controller.close();
                        return;
                    }
                    // Capture current ping time to detect if activity happens during timeout
                    const capturedLastPingTime = lastPingTimeMs;
                    const remainingTimeMs = PING_TIMEOUT_MS - (performance.now() - lastPingTimeMs);
                    // Reuse pending read if it exists (from previous stale timeout iteration)
                    if (!pendingRead) {
                        pendingRead = reader.read();
                    }
                    const currentRead = pendingRead;
                    try {
                        // Race reader.read() against timeout
                        // If timeout fires but activity happened since scheduling, it's "stale" and we retry
                        const result = await Promise.race([
                            currentRead.then((r) => {
                                pendingRead = null;
                                return { type: "data", value: r };
                            }),
                            new Promise((resolve) => setTimeout(() => {
                                if (lastPingTimeMs === capturedLastPingTime) {
                                    // No activity since timeout was scheduled - actually timed out
                                    debug("read session ping timeout");
                                    resolve({ type: "timeout" });
                                }
                                else {
                                    // Activity happened - this timeout is stale, retry with new timeout
                                    debug("stale timeout, activity detected, retrying");
                                    resolve({ type: "stale" });
                                }
                            }, remainingTimeMs)),
                        ]);
                        if (result.type === "stale") {
                            // Loop to create new timeout, reusing pendingRead
                            continue;
                        }
                        if (result.type === "timeout") {
                            const elapsed = performance.now() - lastPingTimeMs;
                            const timeoutError = new error_js_1.S2Error({
                                message: `No ping received for ${Math.floor(elapsed / 1000)}s (timeout: ${PING_TIMEOUT_MS / 1000}s)`,
                                status: 408,
                                code: "TIMEOUT",
                            });
                            controller.enqueue({ ok: false, error: timeoutError });
                            done = true;
                            await reader.cancel();
                            controller.close();
                            return;
                        }
                        // Got data
                        if (result.value.done) {
                            done = true;
                            // Check if stream ended due to error
                            if (parserError) {
                                controller.enqueue({ ok: false, error: parserError });
                            }
                            controller.close();
                        }
                        else {
                            // Emit successful result
                            controller.enqueue({ ok: true, value: result.value.value });
                        }
                        return;
                    }
                    catch (error) {
                        // Convert unexpected errors to S2Error and emit as error result
                        controller.enqueue({ ok: false, error: (0, error_js_1.s2Error)(error) });
                        done = true;
                        await reader.cancel();
                        controller.close();
                        return;
                    }
                }
            },
            cancel: async () => {
                await reader.cancel();
            },
        });
    }
    nextReadPosition() {
        return this._nextReadPosition;
    }
    lastObservedTail() {
        return this._lastObservedTail;
    }
    // Implement AsyncIterable (for await...of support)
    [Symbol.asyncIterator]() {
        const proto = ReadableStream.prototype;
        const fn = proto[Symbol.asyncIterator];
        if (typeof fn === "function") {
            try {
                return fn.call(this);
            }
            catch {
                // Native method may throw "Illegal invocation" when called on subclass
                // Fall through to manual implementation
            }
        }
        const reader = this.getReader();
        return {
            next: async () => {
                const r = await reader.read();
                if (r.done)
                    return { done: true, value: undefined };
                return { done: false, value: r.value };
            },
            return: async (value) => {
                reader.releaseLock();
                return { done: true, value };
            },
            throw: async (e) => {
                reader.releaseLock();
                throw e;
            },
            [Symbol.asyncIterator]() {
                return this;
            },
        };
    }
    // Implement AsyncDisposable
    async [Symbol.asyncDispose]() {
        await this.cancel();
    }
}
exports.FetchReadSession = FetchReadSession;
// Removed AcksStream - transport sessions no longer expose streams
/**
 * Fetch-based transport session for appending records via HTTP/1.1.
 * Queues append requests and ensures only one is in-flight at a time (single-flight).
 * No backpressure, no retry logic, no streams - just submit/close with value-encoded errors.
 */
class FetchAppendSession {
    queue = [];
    pendingResolvers = [];
    inFlight = false;
    options;
    stream;
    closed = false;
    processingPromise = null;
    client;
    static async create(stream, transportConfig, sessionOptions, requestOptions) {
        return new FetchAppendSession(stream, transportConfig, sessionOptions, requestOptions);
    }
    constructor(stream, transportConfig, sessionOptions, requestOptions) {
        this.options = requestOptions;
        this.stream = stream;
        debug("[%s] FetchAppendSession created", stream);
        const headers = {};
        if (transportConfig.basinName) {
            headers["s2-basin"] = transportConfig.basinName;
        }
        if ((0, runtime_js_1.canSetUserAgentHeader)()) {
            headers["user-agent"] = runtime_js_1.DEFAULT_USER_AGENT;
        }
        this.client = (0, index_js_1.createClient)((0, index_js_1.createConfig)({
            baseUrl: transportConfig.baseUrl,
            auth: () => Redacted.value(transportConfig.accessToken),
            headers: headers,
        }));
    }
    /**
     * Close the append session.
     * Waits for all pending appends to complete before resolving.
     * Never throws - returns CloseResult.
     */
    async close() {
        debug("[%s] FetchAppendSession close requested", this.stream);
        try {
            this.closed = true;
            await this.waitForDrain();
            debug("[%s] FetchAppendSession close complete", this.stream);
            return (0, result_js_1.okClose)();
        }
        catch (error) {
            const s2Err = (0, error_js_1.s2Error)(error);
            debug("[%s] FetchAppendSession close error: %s", this.stream, s2Err.message);
            return (0, result_js_1.errClose)(s2Err);
        }
    }
    /**
     * Submit an append request to the session.
     * The request will be queued and sent when no other request is in-flight.
     * Never throws - returns AppendResult discriminated union.
     */
    submit(input) {
        debug("[%s] FetchAppendSession.submit: records=%d, match_seq_num=%s, queueLen=%d", this.stream, input.records.length, input.matchSeqNum ?? "none", this.queue.length);
        // Validate closed state
        if (this.closed) {
            debug("[%s] FetchAppendSession.submit: session closed, rejecting", this.stream);
            return Promise.resolve((0, result_js_1.err)(new error_js_1.S2Error({ message: "AppendSession is closed", status: 400 })));
        }
        // Validate batch size limits (non-retryable 400-level error)
        // Note: This should already be validated by AppendInput.create(), but we check defensively
        if (input.records.length > 1000) {
            debug("[%s] FetchAppendSession.submit: batch too large (%d records)", this.stream, input.records.length);
            return Promise.resolve((0, result_js_1.err)(new error_js_1.S2Error({
                message: `Batch of ${input.records.length} exceeds maximum batch size of 1000 records`,
                status: 400,
                code: "INVALID_ARGUMENT",
            })));
        }
        // Validate metered size (use cached value from AppendInput)
        const batchMeteredSize = input.meteredBytes;
        if (batchMeteredSize > 1024 * 1024) {
            debug("[%s] FetchAppendSession.submit: batch too large (%d bytes)", this.stream, batchMeteredSize);
            return Promise.resolve((0, result_js_1.err)(new error_js_1.S2Error({
                message: `Batch size ${batchMeteredSize} bytes exceeds maximum of 1 MiB (1048576 bytes)`,
                status: 400,
                code: "INVALID_ARGUMENT",
            })));
        }
        return new Promise((resolve) => {
            this.queue.push(input);
            this.pendingResolvers.push({ resolve });
            debug("[%s] FetchAppendSession.submit: queued, queueLen=%d", this.stream, this.queue.length);
            // Start processing if not already running
            if (!this.processingPromise) {
                // Attach a catch to avoid unhandled rejection warnings on hard failures
                this.processingPromise = this.processLoop().catch(() => { });
            }
        });
    }
    /**
     * Main processing loop that sends queued requests one at a time.
     * Single-flight: only one request in progress at a time.
     */
    async processLoop() {
        debug("[%s] FetchAppendSession.processLoop: starting", this.stream);
        while (this.queue.length > 0) {
            this.inFlight = true;
            const input = this.queue.shift();
            const resolver = this.pendingResolvers.shift();
            debug("[%s] FetchAppendSession.processLoop: sending %d records, match_seq_num=%s", this.stream, input.records.length, input.matchSeqNum ?? "none");
            try {
                const preferProtobuf = input.records.some((record) => (0, utils_js_1.computeAppendRecordFormat)(record) === "bytes");
                const appendOptions = this.options
                    ? { ...this.options, preferProtobuf }
                    : { preferProtobuf };
                const ack = await (0, shared_js_1.streamAppend)(this.stream, this.client, input, appendOptions);
                debug("[%s] FetchAppendSession.processLoop: success, seq_num=%d-%d", this.stream, ack.start.seqNum, ack.end.seqNum);
                // Resolve with success result
                resolver.resolve((0, result_js_1.ok)(ack));
            }
            catch (error) {
                // Convert error to S2Error and resolve with error result
                const s2Err = (0, error_js_1.s2Error)(error);
                debug("[%s] FetchAppendSession.processLoop: error, status=%s, message=%s", this.stream, s2Err.status, s2Err.message);
                // Resolve this request with error
                resolver.resolve((0, result_js_1.err)(s2Err));
                // Resolve all remaining pending promises with the same error
                // (transport failure affects all queued requests)
                debug("[%s] FetchAppendSession.processLoop: failing %d queued requests", this.stream, this.pendingResolvers.length);
                for (const pendingResolver of this.pendingResolvers) {
                    pendingResolver.resolve((0, result_js_1.err)(s2Err));
                }
                this.pendingResolvers = [];
                // Clear the queue
                this.queue = [];
                this.inFlight = false;
                this.processingPromise = null;
                return;
            }
            this.inFlight = false;
        }
        debug("[%s] FetchAppendSession.processLoop: done", this.stream);
        this.processingPromise = null;
    }
    async waitForDrain() {
        // Wait for processing to complete
        if (this.processingPromise) {
            await this.processingPromise;
        }
        // Wait until queue is empty and nothing is in flight
        while (this.queue.length > 0 || this.inFlight) {
            await new Promise((resolve) => setTimeout(resolve, 10));
        }
    }
}
exports.FetchAppendSession = FetchAppendSession;
/**
 * Fetch-based transport using HTTP/1.1 + JSON
 * Works in all JavaScript environments (browser, Node.js, Deno, etc.)
 */
class FetchTransport {
    client;
    transportConfig;
    constructor(config) {
        const headers = {};
        if (config.basinName) {
            headers["s2-basin"] = config.basinName;
        }
        if ((0, runtime_js_1.canSetUserAgentHeader)()) {
            headers["user-agent"] = runtime_js_1.DEFAULT_USER_AGENT;
        }
        this.client = (0, index_js_1.createClient)((0, index_js_1.createConfig)({
            baseUrl: config.baseUrl,
            auth: () => Redacted.value(config.accessToken),
            headers: headers,
        }));
        this.transportConfig = config;
    }
    async makeAppendSession(stream, sessionOptions, requestOptions) {
        return retry_js_1.RetryAppendSession.create((myOptions) => {
            return FetchAppendSession.create(stream, this.transportConfig, myOptions, requestOptions);
        }, sessionOptions, this.transportConfig.retry, stream);
    }
    async makeReadSession(stream, args, options) {
        return retry_js_1.RetryReadSession.create((myArgs) => {
            return FetchReadSession.create(this.client, stream, myArgs, options);
        }, args, this.transportConfig.retry);
    }
    async close() {
        // Fetch transport holds no long-lived resources beyond the client.
        // Provided for API symmetry; nothing to tear down.
    }
}
exports.FetchTransport = FetchTransport;
