// @s2-dev/streamstore/dist/cjs/lib/stream/transport/fetch/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.streamRead = streamRead;
exports.streamAppend = streamAppend;
const error_js_1 = require("../../../../error.js");
const index_js_1 = require("../../../../generated/index.js");
const mappers_js_1 = require("../../../../internal/mappers.js");
const utils_js_1 = require("../../../../utils.js");
const proto_js_1 = require("../proto.js");
async function streamRead(stream, client, args, options) {
    const { as, ignore_command_records, ...queryParams } = args ?? {};
    const wantsBytes = (as ?? "string") === "bytes";
    let response;
    try {
        response = await (0, index_js_1.read)({
            client,
            path: {
                stream,
            },
            headers: wantsBytes ? { Accept: "application/protobuf" } : undefined,
            query: queryParams,
            parseAs: wantsBytes ? "arrayBuffer" : undefined,
            ...options,
        });
    }
    catch (error) {
        throw (0, error_js_1.s2Error)(error);
    }
    if (response.error || !response.response.ok) {
        const status = response.response.status;
        if (status === 416) {
            const err = response.error;
            throw new error_js_1.RangeNotSatisfiableError({
                status,
                tail: err?.tail,
                code: err?.code,
            });
        }
        throw (0, error_js_1.makeServerError)({ status, statusText: response.response.statusText }, response.error);
    }
    if (wantsBytes) {
        const batch = (0, proto_js_1.decodeProtoReadBatch)(response.data);
        return batch;
    }
    const res = {
        ...response.data,
        records: response.data.records?.map((record) => ({
            ...record,
            headers: record.headers || undefined,
        })) ?? [],
    };
    return res;
}
async function streamAppend(stream, client, input, options) {
    const { preferProtobuf, ...requestOptions } = options ?? {};
    const hasAnyBytesRecords = input.records.some((record) => (0, utils_js_1.computeAppendRecordFormat)(record) === "bytes");
    const useProtobuf = hasAnyBytesRecords || preferProtobuf === true;
    let response;
    if (useProtobuf) {
        const protoBody = (0, proto_js_1.encodeProtoAppendInput)(input);
        const headers = {
            Accept: "application/protobuf",
            "Content-Type": "application/protobuf",
        };
        try {
            response = await (0, index_js_1.append)({
                client,
                path: {
                    stream,
                },
                body: protoBody,
                bodySerializer: null,
                parseAs: "arrayBuffer",
                headers,
                ...requestOptions,
            });
        }
        catch (error) {
            throw (0, error_js_1.s2Error)(error);
        }
        if (response.error || !response.response.ok) {
            const status = response.response.status;
            if (status === 412) {
                throw (0, error_js_1.makeAppendPreconditionError)(status, response.error);
            }
            throw (0, error_js_1.makeServerError)({ status, statusText: response.response.statusText }, response.error);
        }
        const ack = (0, proto_js_1.decodeProtoAppendAck)(response.data);
        return (0, proto_js_1.protoAppendAckToJson)(ack);
    }
    const encodedRecords = [...input.records].map(mappers_js_1.toAPIAppendRecord);
    try {
        response = await (0, index_js_1.append)({
            client,
            path: {
                stream,
            },
            body: {
                fencing_token: input.fencingToken,
                match_seq_num: input.matchSeqNum,
                records: encodedRecords,
            },
            ...requestOptions,
        });
    }
    catch (error) {
        throw (0, error_js_1.s2Error)(error);
    }
    if (response.error || !response.response.ok) {
        const status = response.response.status;
        if (status === 412) {
            throw (0, error_js_1.makeAppendPreconditionError)(status, response.error);
        }
        throw (0, error_js_1.makeServerError)({ status, statusText: response.response.statusText }, response.error);
    }
    return (0, mappers_js_1.fromAPIAppendAck)(response.data);
}
