// @s2-dev/streamstore/dist/cjs/generated/sdk.gen.js
"use strict";
// This file is auto-generated by @hey-api/openapi-ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTail = exports.append = exports.read = exports.createOrReconfigureStream = exports.reconfigureStream = exports.getStreamConfig = exports.deleteStream = exports.createStream = exports.listStreams = exports.streamMetrics = exports.basinMetrics = exports.accountMetrics = exports.createOrReconfigureBasin = exports.reconfigureBasin = exports.getBasinConfig = exports.deleteBasin = exports.createBasin = exports.listBasins = exports.revokeAccessToken = exports.issueAccessToken = exports.listAccessTokens = void 0;
const client_gen_js_1 = require("./client.gen.js");
/**
 * List access tokens.
 */
const listAccessTokens = (options) => {
    return (options?.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens',
        ...options
    });
};
exports.listAccessTokens = listAccessTokens;
/**
 * Issue a new access token.
 */
const issueAccessToken = (options) => {
    return (options.client ?? client_gen_js_1.client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.issueAccessToken = issueAccessToken;
/**
 * Revoke an access token.
 */
const revokeAccessToken = (options) => {
    return (options.client ?? client_gen_js_1.client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens/{id}',
        ...options
    });
};
exports.revokeAccessToken = revokeAccessToken;
/**
 * List basins.
 */
const listBasins = (options) => {
    return (options?.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins',
        ...options
    });
};
exports.listBasins = listBasins;
/**
 * Create a basin.
 */
const createBasin = (options) => {
    return (options.client ?? client_gen_js_1.client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.createBasin = createBasin;
/**
 * Delete a basin.
 */
const deleteBasin = (options) => {
    return (options.client ?? client_gen_js_1.client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options
    });
};
exports.deleteBasin = deleteBasin;
/**
 * Get basin configuration.
 */
const getBasinConfig = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options
    });
};
exports.getBasinConfig = getBasinConfig;
/**
 * Reconfigure a basin.
 */
const reconfigureBasin = (options) => {
    return (options.client ?? client_gen_js_1.client).patch({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.reconfigureBasin = reconfigureBasin;
/**
 * Create or reconfigure a basin.
 */
const createOrReconfigureBasin = (options) => {
    return (options.client ?? client_gen_js_1.client).put({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.createOrReconfigureBasin = createOrReconfigureBasin;
/**
 * Account-level metrics.
 */
const accountMetrics = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics',
        ...options
    });
};
exports.accountMetrics = accountMetrics;
/**
 * Basin-level metrics.
 */
const basinMetrics = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics/{basin}',
        ...options
    });
};
exports.basinMetrics = basinMetrics;
/**
 * Stream-level metrics.
 */
const streamMetrics = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics/{basin}/{stream}',
        ...options
    });
};
exports.streamMetrics = streamMetrics;
/**
 * List streams.
 */
const listStreams = (options) => {
    return (options?.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams',
        ...options
    });
};
exports.listStreams = listStreams;
/**
 * Create a stream.
 */
const createStream = (options) => {
    return (options.client ?? client_gen_js_1.client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.createStream = createStream;
/**
 * Delete a stream.
 */
const deleteStream = (options) => {
    return (options.client ?? client_gen_js_1.client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options
    });
};
exports.deleteStream = deleteStream;
/**
 * Get stream configuration.
 */
const getStreamConfig = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options
    });
};
exports.getStreamConfig = getStreamConfig;
/**
 * Reconfigure a stream.
 */
const reconfigureStream = (options) => {
    return (options.client ?? client_gen_js_1.client).patch({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.reconfigureStream = reconfigureStream;
/**
 * Create or reconfigure a stream.
 */
const createOrReconfigureStream = (options) => {
    return (options.client ?? client_gen_js_1.client).put({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.createOrReconfigureStream = createOrReconfigureStream;
/**
 * Read records.
 */
const read = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records',
        ...options
    });
};
exports.read = read;
/**
 * Append records.
 */
const append = (options) => {
    return (options.client ?? client_gen_js_1.client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
exports.append = append;
/**
 * Check the tail.
 */
const checkTail = (options) => {
    return (options.client ?? client_gen_js_1.client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records/tail',
        ...options
    });
};
exports.checkTail = checkTail;
