// @trigger.dev/sdk/dist/commonjs/v3/streams.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const streams_js_1 = require("./streams.js");
const v3_1 = require("@trigger.dev/core/v3");
vitest_1.vi.mock("@trigger.dev/core/v3", async (importOriginal) => {
    const original = await importOriginal();
    return {
        ...original,
        taskContext: {
            ctx: {
                run: {
                    id: "run_123",
                    // parentTaskRunId and rootTaskRunId are undefined for root tasks
                },
            },
        },
        realtimeStreams: {
            pipe: vitest_1.vi.fn().mockReturnValue({
                wait: () => Promise.resolve(),
                stream: new ReadableStream(),
            }),
        },
    };
});
(0, vitest_1.describe)("streams.pipe consistency", () => {
    (0, vitest_1.beforeEach)(() => {
        vitest_1.vi.clearAllMocks();
    });
    (0, vitest_1.it)("should not throw and should use self runId when target is 'root' in a root task", async () => {
        const mockStream = new ReadableStream();
        // This should not throw anymore
        const { waitUntilComplete } = streams_js_1.streams.pipe("test-key", mockStream, {
            target: "root",
        });
        (0, vitest_1.expect)(v3_1.realtimeStreams.pipe).toHaveBeenCalledWith("test-key", mockStream, vitest_1.expect.objectContaining({
            target: "run_123",
        }));
    });
    (0, vitest_1.it)("should not throw and should use self runId when target is 'parent' in a root task", async () => {
        const mockStream = new ReadableStream();
        // This should not throw anymore
        const { waitUntilComplete } = streams_js_1.streams.pipe("test-key", mockStream, {
            target: "parent",
        });
        (0, vitest_1.expect)(v3_1.realtimeStreams.pipe).toHaveBeenCalledWith("test-key", mockStream, vitest_1.expect.objectContaining({
            target: "run_123",
        }));
    });
});
