// @s2-dev/streamstore/dist/esm/index.js
// =============================================================================
// Core Client
// =============================================================================
export { S2Basin } from "./basin.js";
export { S2Environment } from "./common.js";
export { S2Endpoints } from "./endpoints.js";
/** Top-level entrypoint for the SDK. */
export { S2 } from "./s2.js";
export { S2Stream } from "./stream.js";
// =============================================================================
// Management Classes
// =============================================================================
export { S2AccessTokens } from "./accessTokens.js";
export { S2Basins } from "./basins.js";
export { S2Metrics } from "./metrics.js";
export { S2Streams } from "./streams.js";
// =============================================================================
// High-Level Helpers
// =============================================================================
/** High-level helper for transforming streams in batches. */
export { BatchTransform } from "./batch-transform.js";
/** High-level producer with automatic batching. */
export { IndexedAppendAck, Producer, RecordSubmitTicket } from "./producer.js";
/**
 * Record types and factory functions for append operations.
 *
 * Use `AppendRecord.string()`, `AppendRecord.bytes()` to create records.
 * Use `AppendInput.create()` to create validated append inputs.
 */
export { AppendInput, AppendRecord, MAX_APPEND_BYTES, MAX_APPEND_RECORDS, } from "./types.js";
/**
 * Generated API types matching the wire format (snake_case field names).
 * Use these when you need to work with raw API data or understand the wire protocol.
 */
export * as API from "./generated/types.gen.js";
// =============================================================================
// Error Types
// =============================================================================
export { FencingTokenMismatchError, RangeNotSatisfiableError, S2Error, SeqNumMismatchError, } from "./error.js";
export { randomToken } from "./lib/base64.js";
export { paginate } from "./lib/paginate.js";
export { meteredBytes, utf8ByteLength } from "./utils.js";
