// @s2-dev/streamstore/dist/esm/stream.js
import { S2Error, withS2Data } from "./error.js";
import { checkTail } from "./generated/index.js";
import { fromAPIReadBatchBytes, fromAPIReadBatchString, fromAPITailResponse, toAPIReadQuery, } from "./internal/mappers.js";
import { withRetries } from "./lib/retry.js";
import { createSessionTransport } from "./lib/stream/factory.js";
import { streamAppend, streamRead, } from "./lib/stream/transport/fetch/shared.js";
import { isCommandRecord } from "./utils.js";
/**
 * Basin-scoped stream helper for append/read operations.
 *
 * Created via {@link S2Basin.stream}. Provides direct methods plus factories for read/append sessions.
 */
export class S2Stream {
    client;
    transportConfig;
    retryConfig;
    _transport;
    closed = false;
    closePromise;
    name;
    constructor(name, client, transportConfig, retryConfig) {
        this.name = name;
        this.client = client;
        this.transportConfig = transportConfig;
        this.retryConfig = retryConfig;
    }
    /**
     * Get or create the transport instance
     */
    async getTransport() {
        this.ensureOpen();
        if (!this._transport) {
            this._transport = await createSessionTransport(this.transportConfig);
        }
        return this._transport;
    }
    ensureOpen() {
        if (this.closed) {
            throw new S2Error({ message: "S2Stream is closed" });
        }
    }
    /**
     * Check the tail of the stream.
     *
     * Returns the next sequence number and timestamp to be assigned (`tail`).
     */
    async checkTail(options) {
        this.ensureOpen();
        return await withRetries(this.retryConfig, async () => {
            const response = await withS2Data(() => checkTail({
                client: this.client,
                path: {
                    stream: this.name,
                },
                ...options,
            }));
            return fromAPITailResponse(response);
        });
    }
    /**
     * Read records from the stream.
     *
     * - When `as: "bytes"` is provided, bodies and headers are decoded from base64 to `Uint8Array`.
     * - Supports starting position by `seq_num`, `timestamp`, or `tail_offset` and can clamp to the tail.
     * - Non-streaming reads are bounded by `count` and `bytes` (defaults 1000 and 1 MiB).
     * - Use `readSession` for streaming reads
     */
    async read(input, options) {
        this.ensureOpen();
        const { as, ...requestOptions } = options ?? {};
        return await withRetries(this.retryConfig, async () => {
            // Convert ReadInput to ReadArgs using mapper
            const readArgs = {
                ...toAPIReadQuery(input),
                as,
            };
            const genBatch = await streamRead(this.name, this.client, readArgs, requestOptions);
            // Convert from API to SDK ReadBatch
            const batch = (as === "bytes"
                ? fromAPIReadBatchBytes(genBatch)
                : fromAPIReadBatchString(genBatch));
            if (input?.ignoreCommandRecords) {
                return {
                    ...batch,
                    records: batch.records.filter((r) => !isCommandRecord(r)),
                };
            }
            return batch;
        });
    }
    /**
     * Append a batch of records to the stream.
     *
     * - Automatically base64-encodes when format is "bytes".
     * - Supports conditional appends via `fencingToken` and `matchSeqNum` in the input.
     * - Returns the acknowledged range and the stream tail after the append.
     * - All records in a batch must use the same format (either all string or all bytes).
     *
     * Use {@link AppendInput.create} to construct a validated AppendInput.
     * For high-throughput sequential appends, use `appendSession()` instead.
     *
     * @param input The append input containing records and optional conditions
     * @param options Optional request options
     */
    async append(input, options) {
        this.ensureOpen();
        return await withRetries(this.retryConfig, async () => {
            return await streamAppend(this.name, this.client, input, options);
        }, (config, error) => {
            if ((config.appendRetryPolicy ?? "all") === "noSideEffects") {
                // Allow retry only when the append is naturally idempotent by containing
                // a match_seq_num condition.
                return input.matchSeqNum !== undefined;
            }
            else {
                return true;
            }
        });
    }
    /**
     * Open a streaming read session
     *
     * Use the returned session as an async iterable or as a readable stream.
     * When `as: "bytes"` is provided, bodies and headers are decoded to `Uint8Array`.
     */
    async readSession(input, options) {
        this.ensureOpen();
        const { as, ...requestOptions } = options ?? {};
        const transport = await this.getTransport();
        // Convert ReadInput to ReadArgs using mapper
        const readArgs = {
            ...toAPIReadQuery(input),
            as,
            ignore_command_records: input?.ignoreCommandRecords,
        };
        return await transport.makeReadSession(this.name, readArgs, requestOptions);
    }
    /**
     * Create an append session that guarantees ordering of submissions.
     *
     * Use this to coordinate high-throughput, sequential appends with backpressure.
     * Records can be either string or bytes format - the format is specified in each record.
     *
     * @param sessionOptions Options that control append session behavior
     * @param requestOptions Optional request options
     */
    async appendSession(sessionOptions, requestOptions) {
        this.ensureOpen();
        const transport = await this.getTransport();
        return await transport.makeAppendSession(this.name, sessionOptions, requestOptions);
    }
    async close() {
        if (this.closePromise) {
            return this.closePromise;
        }
        this.closePromise = (async () => {
            if (this.closed) {
                return;
            }
            this.closed = true;
            if (this._transport) {
                try {
                    await this._transport.close();
                }
                finally {
                    this._transport = undefined;
                }
            }
        })();
        try {
            await this.closePromise;
        }
        finally {
            this.closePromise = undefined;
        }
    }
    async [Symbol.asyncDispose]() {
        await this.close();
    }
}
