// @s2-dev/streamstore/dist/esm/generated/sdk.gen.js
// This file is auto-generated by @hey-api/openapi-ts
import { client } from './client.gen.js';
/**
 * List access tokens.
 */
export const listAccessTokens = (options) => {
    return (options?.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens',
        ...options
    });
};
/**
 * Issue a new access token.
 */
export const issueAccessToken = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Revoke an access token.
 */
export const revokeAccessToken = (options) => {
    return (options.client ?? client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/access-tokens/{id}',
        ...options
    });
};
/**
 * List basins.
 */
export const listBasins = (options) => {
    return (options?.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins',
        ...options
    });
};
/**
 * Create a basin.
 */
export const createBasin = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Delete a basin.
 */
export const deleteBasin = (options) => {
    return (options.client ?? client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options
    });
};
/**
 * Get basin configuration.
 */
export const getBasinConfig = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options
    });
};
/**
 * Reconfigure a basin.
 */
export const reconfigureBasin = (options) => {
    return (options.client ?? client).patch({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Create or reconfigure a basin.
 */
export const createOrReconfigureBasin = (options) => {
    return (options.client ?? client).put({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/basins/{basin}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Account-level metrics.
 */
export const accountMetrics = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics',
        ...options
    });
};
/**
 * Basin-level metrics.
 */
export const basinMetrics = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics/{basin}',
        ...options
    });
};
/**
 * Stream-level metrics.
 */
export const streamMetrics = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/metrics/{basin}/{stream}',
        ...options
    });
};
/**
 * List streams.
 */
export const listStreams = (options) => {
    return (options?.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams',
        ...options
    });
};
/**
 * Create a stream.
 */
export const createStream = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Delete a stream.
 */
export const deleteStream = (options) => {
    return (options.client ?? client).delete({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options
    });
};
/**
 * Get stream configuration.
 */
export const getStreamConfig = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options
    });
};
/**
 * Reconfigure a stream.
 */
export const reconfigureStream = (options) => {
    return (options.client ?? client).patch({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Create or reconfigure a stream.
 */
export const createOrReconfigureStream = (options) => {
    return (options.client ?? client).put({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Read records.
 */
export const read = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records',
        ...options
    });
};
/**
 * Append records.
 */
export const append = (options) => {
    return (options.client ?? client).post({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records',
        ...options,
        headers: {
            'Content-Type': 'application/json',
            ...options.headers
        }
    });
};
/**
 * Check the tail.
 */
export const checkTail = (options) => {
    return (options.client ?? client).get({
        security: [
            {
                scheme: 'bearer',
                type: 'http'
            }
        ],
        url: '/streams/{stream}/records/tail',
        ...options
    });
};
