// @trigger.dev/core/dist/commonjs/v3/otel/diskIoMetrics.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startDiskIoMetrics = startDiskIoMetrics;
const fs = __importStar(require("node:fs"));
const fsPromises = __importStar(require("node:fs/promises"));
const SECTOR_SIZE = 512;
const FILTERED_DEVICE_PREFIXES = ["loop", "ram", "dm-"];
function parseProcDiskstats(content) {
    const entries = [];
    for (const line of content.split("\n")) {
        const trimmed = line.trim();
        if (!trimmed)
            continue;
        const fields = trimmed.split(/\s+/);
        if (fields.length < 14)
            continue;
        const device = fields[2];
        if (FILTERED_DEVICE_PREFIXES.some((prefix) => device.startsWith(prefix))) {
            continue;
        }
        entries.push({
            device,
            readsCompleted: parseInt(fields[3], 10),
            sectorsRead: parseInt(fields[5], 10),
            writesCompleted: parseInt(fields[7], 10),
            sectorsWritten: parseInt(fields[9], 10),
        });
    }
    return entries;
}
function startDiskIoMetrics(meterProvider) {
    try {
        fs.accessSync("/proc/diskstats", fs.constants.R_OK);
    }
    catch {
        return;
    }
    const meter = meterProvider.getMeter("system-disk", "1.0.0");
    const ioCounter = meter.createObservableCounter("system.disk.io", {
        description: "Disk I/O bytes read and written per device",
        unit: "By",
    });
    const opsCounter = meter.createObservableCounter("system.disk.operations", {
        description: "Disk read/write operation counts per device",
        unit: "{operation}",
    });
    meter.addBatchObservableCallback(async (obs) => {
        try {
            const content = await fsPromises.readFile("/proc/diskstats", "utf-8");
            const stats = parseProcDiskstats(content);
            for (const entry of stats) {
                const readAttrs = {
                    "system.device": entry.device,
                    "disk.io.direction": "read",
                };
                const writeAttrs = {
                    "system.device": entry.device,
                    "disk.io.direction": "write",
                };
                obs.observe(ioCounter, entry.sectorsRead * SECTOR_SIZE, readAttrs);
                obs.observe(ioCounter, entry.sectorsWritten * SECTOR_SIZE, writeAttrs);
                obs.observe(opsCounter, entry.readsCompleted, readAttrs);
                obs.observe(opsCounter, entry.writesCompleted, writeAttrs);
            }
        }
        catch {
            // Skip entire cycle on failure
        }
    }, [ioCounter, opsCounter]);
}
