// @s2-dev/streamstore/dist/cjs/basins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2Basins = void 0;
const error_js_1 = require("./error.js");
const index_js_1 = require("./generated/index.js");
const case_transform_js_1 = require("./internal/case-transform.js");
const base64_js_1 = require("./lib/base64.js");
const paginate_js_1 = require("./lib/paginate.js");
const retry_js_1 = require("./lib/retry.js");
/** Convert SDK RetentionPolicy (ageSecs) to API RetentionPolicy (age). */
function toAPIRetentionPolicy(policy) {
    if (policy === null)
        return null;
    if (policy === undefined)
        return undefined;
    if ("ageSecs" in policy) {
        return { age: Math.floor(policy.ageSecs) };
    }
    return policy; // { infinite: ... } passes through
}
/** Convert API RetentionPolicy (age) to SDK RetentionPolicy (ageSecs). */
function toSDKRetentionPolicy(policy) {
    if (policy === null)
        return null;
    if (policy === undefined)
        return undefined;
    if ("age" in policy) {
        return { ageSecs: policy.age };
    }
    return policy; // { infinite: ... } passes through
}
/** Convert SDK StreamConfig to API format (handles retentionPolicy.ageSecs → age). */
function toAPIStreamConfig(config) {
    if (config === null || config === undefined)
        return config;
    return {
        ...config,
        deleteOnEmpty: config.deleteOnEmpty
            ? {
                ...config.deleteOnEmpty,
                minAgeSecs: config.deleteOnEmpty.minAgeSecs === undefined
                    ? undefined
                    : Math.max(0, Math.floor(config.deleteOnEmpty.minAgeSecs)),
            }
            : config.deleteOnEmpty,
        retentionPolicy: toAPIRetentionPolicy(config.retentionPolicy),
    };
}
/** Convert API StreamConfig to SDK format (handles retentionPolicy.age → ageSecs). */
function toSDKStreamConfig(config) {
    if (config === null || config === undefined)
        return config;
    return {
        ...config,
        retentionPolicy: toSDKRetentionPolicy(config?.retentionPolicy),
    };
}
/** Convert SDK BasinConfig to API format. */
function toAPIBasinConfig(config) {
    if (config === null || config === undefined)
        return config;
    return {
        ...config,
        defaultStreamConfig: toAPIStreamConfig(config.defaultStreamConfig),
    };
}
/** Convert API BasinConfig to SDK format. */
function toSDKBasinConfig(config) {
    return {
        ...config,
        defaultStreamConfig: toSDKStreamConfig(config?.defaultStreamConfig),
    };
}
/**
 * Account-scoped helper for listing, creating, deleting, and reconfiguring basins.
 *
 * Retrieve this via {@link S2.basins}. Each method retries according to the client-level retry config.
 */
class S2Basins {
    client;
    retryConfig;
    constructor(client, retryConfig) {
        this.client = client;
        this.retryConfig = retryConfig;
    }
    /**
     * List basins.
     *
     * @param args.prefix Return basins whose names start with the given prefix
     * @param args.startAfter Name to start after (for pagination)
     * @param args.limit Max results (up to 1000)
     */
    async list(args, options) {
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.listBasins)({
                client: this.client,
                query: (0, case_transform_js_1.toSnakeCase)(args),
                ...options,
            }));
        });
        return (0, case_transform_js_1.toCamelCase)(response);
    }
    /**
     * List all basins with automatic pagination.
     * Returns a lazy async iterable that fetches pages as needed.
     *
     * @param args - Optional options: `prefix` to filter by name prefix, `limit` for max results per page, `includeDeleted` to include basins pending deletion
     *
     * @example
     * ```ts
     * for await (const basin of s2.basins.listAll({ prefix: "my-" })) {
     *   console.log(basin.name);
     * }
     * ```
     */
    listAll(args, options) {
        const { includeDeleted, ...listArgs } = args ?? {};
        const allItems = (0, paginate_js_1.paginate)((a) => this.list(a, options).then((r) => ({
            items: r.basins,
            hasMore: r.hasMore,
        })), listArgs, (basin) => basin.name);
        if (includeDeleted) {
            return allItems;
        }
        return (0, paginate_js_1.filterAsync)(allItems, (b) => b.state !== "deleting");
    }
    /**
     * Create a basin.
     *
     * @param args.basin Globally unique basin name (8-48 chars, lowercase letters, numbers, hyphens; cannot begin or end with a hyphen)
     * @param args.config Optional basin configuration (e.g. defaultStreamConfig)
     * @param args.scope Basin scope
     */
    async create(args, options) {
        const requestToken = (0, base64_js_1.randomToken)();
        // Convert SDK config to API format (ageSecs → age)
        const apiArgs = {
            ...args,
            config: toAPIBasinConfig(args.config),
        };
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.createBasin)({
                client: this.client,
                body: (0, case_transform_js_1.toSnakeCase)(apiArgs),
                headers: { "s2-request-token": requestToken },
                ...options,
            }));
        });
        return (0, case_transform_js_1.toCamelCase)(response);
    }
    /**
     * Get basin configuration.
     *
     * @param args.basin Basin name
     */
    async getConfig(args, options) {
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.getBasinConfig)({
                client: this.client,
                path: args,
                ...options,
            }));
        });
        // Convert API format to SDK (age → ageSecs)
        return toSDKBasinConfig((0, case_transform_js_1.toCamelCase)(response));
    }
    /**
     * Delete a basin.
     *
     * @param args.basin Basin name
     */
    async delete(args, options) {
        await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.deleteBasin)({
                client: this.client,
                path: args,
                ...options,
            }));
        });
    }
    /**
     * Reconfigure a basin.
     *
     * @param args Configuration for the basin to reconfigure (including basin name and fields to change)
     */
    async reconfigure(args, options) {
        // Convert SDK config to API format (ageSecs → age)
        const apiArgs = {
            ...args,
            defaultStreamConfig: toAPIStreamConfig(args.defaultStreamConfig),
        };
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.reconfigureBasin)({
                client: this.client,
                path: args,
                body: (0, case_transform_js_1.toSnakeCase)(apiArgs),
                ...options,
            }));
        });
        // Convert API format to SDK (age → ageSecs)
        return toSDKBasinConfig((0, case_transform_js_1.toCamelCase)(response));
    }
}
exports.S2Basins = S2Basins;
