// @s2-dev/streamstore/dist/cjs/basin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2Basin = void 0;
const index_js_1 = require("./generated/client/index.js");
const Redacted = require("./lib/redacted.js");
const runtime_js_1 = require("./lib/stream/runtime.js");
const stream_js_1 = require("./stream.js");
const streams_js_1 = require("./streams.js");
class S2Basin {
    client;
    transportConfig;
    retryConfig;
    name;
    streams;
    /**
     * Create a basin-scoped client that talks to `https://{basin}.b.aws.s2.dev/v1`.
     *
     * Use this to work with streams inside a single basin.
     * @param name Basin name
     * @param options Configuration for the basin-scoped client
     */
    constructor(name, options) {
        this.name = name;
        this.retryConfig = options.retryConfig;
        this.transportConfig = {
            baseUrl: options.baseUrl,
            accessToken: options.accessToken,
            basinName: options.includeBasinHeader ? name : undefined,
            connectionTimeoutMillis: options.retryConfig?.connectionTimeoutMillis,
            requestTimeoutMillis: options.retryConfig?.requestTimeoutMillis,
            retry: options.retryConfig,
        };
        const headers = {};
        if (options.includeBasinHeader) {
            headers["s2-basin"] = name;
        }
        if ((0, runtime_js_1.canSetUserAgentHeader)()) {
            headers["user-agent"] = runtime_js_1.DEFAULT_USER_AGENT;
        }
        this.client = (0, index_js_1.createClient)((0, index_js_1.createConfig)({
            baseUrl: options.baseUrl,
            auth: () => Redacted.value(this.transportConfig.accessToken),
            headers: headers,
        }));
        this.streams = new streams_js_1.S2Streams(this.client, this.retryConfig);
    }
    /**
     * Create a stream-scoped helper bound to `this` basin.
     * @param name Stream name
     */
    stream(name, options) {
        return new stream_js_1.S2Stream(name, this.client, {
            ...this.transportConfig,
            forceTransport: options?.forceTransport,
        }, this.retryConfig);
    }
}
exports.S2Basin = S2Basin;
