// @s2-dev/streamstore/dist/cjs/index.js
"use strict";
// =============================================================================
// Core Client
// =============================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.utf8ByteLength = exports.meteredBytes = exports.paginate = exports.randomToken = exports.SeqNumMismatchError = exports.S2Error = exports.RangeNotSatisfiableError = exports.FencingTokenMismatchError = exports.API = exports.MAX_APPEND_RECORDS = exports.MAX_APPEND_BYTES = exports.AppendRecord = exports.AppendInput = exports.RecordSubmitTicket = exports.Producer = exports.IndexedAppendAck = exports.BatchTransform = exports.S2Streams = exports.S2Metrics = exports.S2Basins = exports.S2AccessTokens = exports.S2Stream = exports.S2 = exports.S2Endpoints = exports.S2Environment = exports.S2Basin = void 0;
var basin_js_1 = require("./basin.js");
Object.defineProperty(exports, "S2Basin", { enumerable: true, get: function () { return basin_js_1.S2Basin; } });
var common_js_1 = require("./common.js");
Object.defineProperty(exports, "S2Environment", { enumerable: true, get: function () { return common_js_1.S2Environment; } });
var endpoints_js_1 = require("./endpoints.js");
Object.defineProperty(exports, "S2Endpoints", { enumerable: true, get: function () { return endpoints_js_1.S2Endpoints; } });
/** Top-level entrypoint for the SDK. */
var s2_js_1 = require("./s2.js");
Object.defineProperty(exports, "S2", { enumerable: true, get: function () { return s2_js_1.S2; } });
var stream_js_1 = require("./stream.js");
Object.defineProperty(exports, "S2Stream", { enumerable: true, get: function () { return stream_js_1.S2Stream; } });
// =============================================================================
// Management Classes
// =============================================================================
var accessTokens_js_1 = require("./accessTokens.js");
Object.defineProperty(exports, "S2AccessTokens", { enumerable: true, get: function () { return accessTokens_js_1.S2AccessTokens; } });
var basins_js_1 = require("./basins.js");
Object.defineProperty(exports, "S2Basins", { enumerable: true, get: function () { return basins_js_1.S2Basins; } });
var metrics_js_1 = require("./metrics.js");
Object.defineProperty(exports, "S2Metrics", { enumerable: true, get: function () { return metrics_js_1.S2Metrics; } });
var streams_js_1 = require("./streams.js");
Object.defineProperty(exports, "S2Streams", { enumerable: true, get: function () { return streams_js_1.S2Streams; } });
// =============================================================================
// High-Level Helpers
// =============================================================================
/** High-level helper for transforming streams in batches. */
var batch_transform_js_1 = require("./batch-transform.js");
Object.defineProperty(exports, "BatchTransform", { enumerable: true, get: function () { return batch_transform_js_1.BatchTransform; } });
/** High-level producer with automatic batching. */
var producer_js_1 = require("./producer.js");
Object.defineProperty(exports, "IndexedAppendAck", { enumerable: true, get: function () { return producer_js_1.IndexedAppendAck; } });
Object.defineProperty(exports, "Producer", { enumerable: true, get: function () { return producer_js_1.Producer; } });
Object.defineProperty(exports, "RecordSubmitTicket", { enumerable: true, get: function () { return producer_js_1.RecordSubmitTicket; } });
/**
 * Record types and factory functions for append operations.
 *
 * Use `AppendRecord.string()`, `AppendRecord.bytes()` to create records.
 * Use `AppendInput.create()` to create validated append inputs.
 */
var types_js_1 = require("./types.js");
Object.defineProperty(exports, "AppendInput", { enumerable: true, get: function () { return types_js_1.AppendInput; } });
Object.defineProperty(exports, "AppendRecord", { enumerable: true, get: function () { return types_js_1.AppendRecord; } });
Object.defineProperty(exports, "MAX_APPEND_BYTES", { enumerable: true, get: function () { return types_js_1.MAX_APPEND_BYTES; } });
Object.defineProperty(exports, "MAX_APPEND_RECORDS", { enumerable: true, get: function () { return types_js_1.MAX_APPEND_RECORDS; } });
/**
 * Generated API types matching the wire format (snake_case field names).
 * Use these when you need to work with raw API data or understand the wire protocol.
 */
exports.API = require("./generated/types.gen.js");
// =============================================================================
// Error Types
// =============================================================================
var error_js_1 = require("./error.js");
Object.defineProperty(exports, "FencingTokenMismatchError", { enumerable: true, get: function () { return error_js_1.FencingTokenMismatchError; } });
Object.defineProperty(exports, "RangeNotSatisfiableError", { enumerable: true, get: function () { return error_js_1.RangeNotSatisfiableError; } });
Object.defineProperty(exports, "S2Error", { enumerable: true, get: function () { return error_js_1.S2Error; } });
Object.defineProperty(exports, "SeqNumMismatchError", { enumerable: true, get: function () { return error_js_1.SeqNumMismatchError; } });
var base64_js_1 = require("./lib/base64.js");
Object.defineProperty(exports, "randomToken", { enumerable: true, get: function () { return base64_js_1.randomToken; } });
var paginate_js_1 = require("./lib/paginate.js");
Object.defineProperty(exports, "paginate", { enumerable: true, get: function () { return paginate_js_1.paginate; } });
var utils_js_1 = require("./utils.js");
Object.defineProperty(exports, "meteredBytes", { enumerable: true, get: function () { return utils_js_1.meteredBytes; } });
Object.defineProperty(exports, "utf8ByteLength", { enumerable: true, get: function () { return utils_js_1.utf8ByteLength; } });
