// @s2-dev/streamstore/dist/cjs/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2Metrics = void 0;
exports.fromAPIMetricSetResponse = fromAPIMetricSetResponse;
const error_js_1 = require("./error.js");
const index_js_1 = require("./generated/index.js");
const case_transform_js_1 = require("./internal/case-transform.js");
const retry_js_1 = require("./lib/retry.js");
/** Convert timestamp (Date or milliseconds) to Unix seconds (floored). */
function toEpochSeconds(value) {
    if (value === undefined)
        return undefined;
    const ms = typeof value === "number" ? value : value.getTime();
    return Math.floor(ms / 1000);
}
/** Convert API metric response to SDK types with Date conversions. */
function fromAPIMetricSetResponse(response) {
    const camelCased = (0, case_transform_js_1.toCamelCase)(response);
    // Convert timeseries timestamps from seconds to Date
    return {
        values: camelCased.values.map((metric) => {
            if ("accumulation" in metric) {
                return {
                    accumulation: {
                        ...metric.accumulation,
                        values: metric.accumulation.values.map(([ts, value]) => [new Date(ts * 1000), value]),
                    },
                };
            }
            if ("gauge" in metric) {
                return {
                    gauge: {
                        ...metric.gauge,
                        values: metric.gauge.values.map(([ts, value]) => [new Date(ts * 1000), value]),
                    },
                };
            }
            return metric;
        }),
    };
}
/**
 * Helper for querying account, basin, and stream level metrics.
 *
 * Access via {@link S2.metrics}. Responses are automatically converted to Date-friendly SDK types.
 */
class S2Metrics {
    client;
    retryConfig;
    constructor(client, retryConfig) {
        this.client = client;
        this.retryConfig = retryConfig;
    }
    /**
     * Account-level metrics.
     *
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (milliseconds since Unix epoch)
     * @param args.end Optional end timestamp (milliseconds since Unix epoch)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async account(args, options) {
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.accountMetrics)({
                client: this.client,
                query: (0, case_transform_js_1.toSnakeCase)({
                    ...args,
                    start: toEpochSeconds(args.start),
                    end: toEpochSeconds(args.end),
                }),
                ...options,
            }));
        });
        return fromAPIMetricSetResponse(response);
    }
    /**
     * Basin-level metrics.
     *
     * @param args.basin Basin name
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (milliseconds since Unix epoch)
     * @param args.end Optional end timestamp (milliseconds since Unix epoch)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async basin(args, options) {
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.basinMetrics)({
                client: this.client,
                path: args,
                query: (0, case_transform_js_1.toSnakeCase)({
                    ...args,
                    start: toEpochSeconds(args.start),
                    end: toEpochSeconds(args.end),
                }),
                ...options,
            }));
        });
        return fromAPIMetricSetResponse(response);
    }
    /**
     * Stream-level metrics.
     *
     * @param args.basin Basin name
     * @param args.stream Stream name
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (milliseconds since Unix epoch)
     * @param args.end Optional end timestamp (milliseconds since Unix epoch)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async stream(args, options) {
        const response = await (0, retry_js_1.withRetries)(this.retryConfig, async () => {
            return await (0, error_js_1.withS2Data)(() => (0, index_js_1.streamMetrics)({
                client: this.client,
                path: args,
                query: (0, case_transform_js_1.toSnakeCase)({
                    ...args,
                    start: toEpochSeconds(args.start),
                    end: toEpochSeconds(args.end),
                }),
                ...options,
            }));
        });
        return fromAPIMetricSetResponse(response);
    }
}
exports.S2Metrics = S2Metrics;
