// @s2-dev/streamstore/dist/cjs/producer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Producer = exports.RecordSubmitTicket = exports.IndexedAppendAck = void 0;
const debug_1 = require("debug");
const error_js_1 = require("./error.js");
const types_js_1 = require("./types.js");
const debugProducer = (0, debug_1.default)("s2:producer");
const toS2Error = (err) => err instanceof error_js_1.S2Error
    ? err
    : new error_js_1.S2Error({
        message: String(err),
        status: 500,
        origin: "sdk",
    });
class IndexedAppendAck {
    index;
    ack;
    constructor(index, ack) {
        this.index = index;
        this.ack = ack;
    }
    batchAppendAck() {
        return this.ack;
    }
    seqNum() {
        return this.ack.start.seqNum + this.index;
    }
}
exports.IndexedAppendAck = IndexedAppendAck;
class RecordSubmitTicket {
    ackPromise;
    constructor(ackPromise) {
        this.ackPromise = ackPromise;
        // Avoid unhandled rejections if the caller never awaits ack().
        this.ackPromise.catch(() => { });
    }
    /**
     * Returns a promise that resolves with the IndexedAppendAck once the record is durable.
     */
    ack() {
        return this.ackPromise;
    }
}
exports.RecordSubmitTicket = RecordSubmitTicket;
/**
 * Producer provides per-record append semantics on top of a batched AppendSession.
 *
 * - submit(record) returns a Promise<RecordSubmitTicket> that resolves once the record
 *   has been accepted (written to the batch transform). Backpressure is applied
 *   automatically via the transform stream when the AppendSession is at capacity.
 * - ticket.ack() returns a Promise<IndexedAppendAck> that resolves once the record is durable.
 *
 * See the "Producer API" section of the root README for guidance on sizing batches,
 * wiring transforms, and handling application-level ids.
 *
 * @example
 * ```ts
 * const appendSession = await stream.appendSession();
 * const producer = new Producer(new BatchTransform(), appendSession);
 * const writer = producer.writable.getWriter();
 * await writer.write(AppendRecord.string({ body: "hello" }));
 * await writer.close();
 *
 * for await (const ack of producer.readable) {
 *   console.log("record durable at seq", ack.seqNum());
 * }
 *
 * await producer.close();
 * ```
 */
class Producer {
    batchTransform;
    transformWriter;
    transformReader;
    pump;
    appendSession;
    readable;
    writable;
    inflightRecords = [];
    pumpError = null;
    readableController = null;
    debugName;
    submitCounter = 0;
    constructor(batchTransform, appendSession, debugName) {
        this.debugName = debugName ?? `producer-${Date.now()}`;
        this.batchTransform = batchTransform;
        this.transformWriter = batchTransform.writable.getWriter();
        this.transformReader = batchTransform.readable.getReader();
        this.appendSession = appendSession;
        debugProducer("[%s] created", this.debugName);
        // Create readable stream that emits individual record acknowledgements
        this.readable = new ReadableStream({
            start: (controller) => {
                this.readableController = controller;
            },
            cancel: () => {
                this.close().catch(() => {
                    // Ignore errors during cleanup
                });
            },
        });
        // Create writable stream that accepts individual records
        this.writable = new WritableStream({
            write: async (record) => {
                await this.submit(record);
            },
            close: async () => {
                await this.close();
            },
            abort: async (reason) => {
                this.pumpError = toS2Error(reason);
                await this.close();
            },
        });
        this.pump = this.runPump();
    }
    /**
     * Main pump loop: reads batches from transform, submits to session, handles acks.
     */
    async runPump() {
        debugProducer("[%s] pump started", this.debugName);
        try {
            while (true) {
                const { value: batch, done } = await this.transformReader.read();
                if (done) {
                    debugProducer("[%s] pump done (transform closed)", this.debugName);
                    break;
                }
                debugProducer("[%s] pump got batch: records=%d, match_seq_num=%s, inflightRecords=%d", this.debugName, batch.records.length, batch.matchSeqNum ?? "none", this.inflightRecords.length);
                // Associate records with this batch (FIFO correspondence)
                const recordCount = batch.records.length;
                const associatedRecords = this.inflightRecords.splice(0, recordCount);
                if (associatedRecords.length !== recordCount) {
                    throw new error_js_1.S2Error({
                        message: `Internal error: flushed ${recordCount} records but only ${associatedRecords.length} inflight entries`,
                        status: 500,
                        origin: "sdk",
                    });
                }
                // Submit to AppendSession (blocks on capacity)
                let ticket;
                try {
                    debugProducer("[%s] pump submitting to session: records=%d, match_seq_num=%s", this.debugName, batch.records.length, batch.matchSeqNum ?? "none");
                    const input = types_js_1.AppendInput.create(batch.records, {
                        fencingToken: batch.fencingToken,
                        matchSeqNum: batch.matchSeqNum,
                    });
                    ticket = await this.appendSession.submit(input);
                    debugProducer("[%s] pump submit returned ticket", this.debugName);
                }
                catch (err) {
                    const error = toS2Error(err);
                    debugProducer("[%s] pump submit error: %s", this.debugName, error.message);
                    if (!this.pumpError) {
                        this.pumpError = error;
                    }
                    // Reject acks for records in this batch
                    for (const record of associatedRecords) {
                        record.rejectAck(error);
                    }
                    if (this.readableController) {
                        this.readableController.error(error);
                    }
                    continue;
                }
                // Handle ack asynchronously (non-blocking)
                ticket
                    .ack()
                    .then((ack) => {
                    debugProducer("[%s] pump ack received: seq_num=%d-%d", this.debugName, ack.start.seqNum, ack.end.seqNum);
                    for (const [i, record] of associatedRecords.entries()) {
                        const indexedAck = new IndexedAppendAck(i, ack);
                        record.resolveAck(indexedAck);
                        if (this.readableController) {
                            this.readableController.enqueue(indexedAck);
                        }
                    }
                })
                    .catch((err) => {
                    const error = toS2Error(err);
                    debugProducer("[%s] pump ack error: %s", this.debugName, error.message);
                    if (!this.pumpError) {
                        this.pumpError = error;
                    }
                    for (const record of associatedRecords) {
                        record.rejectAck(error);
                    }
                    if (this.readableController) {
                        this.readableController.error(error);
                    }
                });
            }
        }
        catch (err) {
            const error = toS2Error(err);
            debugProducer("[%s] pump caught error: %s", this.debugName, error.message);
            if (!this.pumpError) {
                this.pumpError = error;
            }
            // Reject all remaining inflight records
            for (const record of this.inflightRecords.splice(0)) {
                record.rejectAck(error);
            }
            // Error the readable stream
            if (this.readableController) {
                this.readableController.error(error);
                this.readableController = null;
            }
        }
    }
    /**
     * Submit a single record for appending.
     *
     * Returns a promise that resolves to a RecordSubmitTicket once the record has been
     * accepted. The promise blocks if the underlying AppendSession is at capacity
     * (backpressure is applied via the transform stream).
     *
     * @throws S2Error if the Producer has failed
     */
    async submit(record) {
        const submitId = ++this.submitCounter;
        debugProducer("[%s] submit #%d: inflightRecords=%d", this.debugName, submitId, this.inflightRecords.length);
        // Check if pump has already failed
        if (this.pumpError) {
            debugProducer("[%s] submit #%d: pump already failed", this.debugName, submitId);
            throw new error_js_1.S2Error({
                message: `Cannot submit: producer has failed: ${this.pumpError.message}`,
                status: 500,
                origin: "sdk",
            });
        }
        // Create the ack promise (resolved later by pump)
        let resolveAck;
        let rejectAck;
        const ackPromise = new Promise((resolve, reject) => {
            resolveAck = resolve;
            rejectAck = reject;
        });
        // Suppress unhandled rejection if write fails before we return the ticket
        ackPromise.catch(() => { });
        // Track this record
        const entry = { resolveAck, rejectAck };
        this.inflightRecords.push(entry);
        debugProducer("[%s] submit #%d: pushed to inflightRecords (now %d), writing to transform", this.debugName, submitId, this.inflightRecords.length);
        try {
            // Write to transform - BLOCKS on backpressure
            await this.transformWriter.write(record);
            debugProducer("[%s] submit #%d: write completed", this.debugName, submitId);
        }
        catch (err) {
            debugProducer("[%s] submit #%d: write failed: %s", this.debugName, submitId, err);
            // Remove from inflight if the write failed
            const idx = this.inflightRecords.indexOf(entry);
            if (idx >= 0) {
                this.inflightRecords.splice(idx, 1);
            }
            const error = toS2Error(err);
            rejectAck(error);
            throw error;
        }
        // Write succeeded - return ticket immediately
        return new RecordSubmitTicket(ackPromise);
    }
    /**
     * Close the Producer gracefully.
     *
     * Waits for all pending records to be flushed, submitted, and acknowledged.
     * If any error occurred during the Producer's lifetime, this method throws it.
     */
    async close() {
        debugProducer("[%s] close requested", this.debugName);
        // Close the writer to signal no more records
        await this.transformWriter.close();
        // Wait for the pump to finish processing all batches
        await this.pump;
        // Close the underlying session
        await this.appendSession.close();
        // Reject any remaining inflight records (shouldn't happen in normal operation)
        if (this.inflightRecords.length > 0) {
            const closingError = new error_js_1.S2Error({
                message: "Producer closed with pending records",
                status: 499,
                origin: "sdk",
            });
            for (const record of this.inflightRecords.splice(0)) {
                record.rejectAck(closingError);
            }
        }
        // Close the readable stream
        if (this.readableController) {
            this.readableController.close();
            this.readableController = null;
        }
        debugProducer("[%s] close complete", this.debugName);
        // If an error occurred, throw it
        if (this.pumpError) {
            throw this.pumpError;
        }
    }
    async [Symbol.asyncDispose]() {
        await this.close();
    }
}
exports.Producer = Producer;
