// @s2-dev/streamstore/dist/cjs/s2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2 = void 0;
const accessTokens_js_1 = require("./accessTokens.js");
const basin_js_1 = require("./basin.js");
const basins_js_1 = require("./basins.js");
const endpoints_js_1 = require("./endpoints.js");
const error_js_1 = require("./error.js");
const index_js_1 = require("./generated/client/index.js");
const Redacted = require("./lib/redacted.js");
const runtime_js_1 = require("./lib/stream/runtime.js");
const metrics_js_1 = require("./metrics.js");
/**
 * Basin names must be 8-48 characters, lowercase alphanumeric and hyphens,
 * cannot start or end with a hyphen.
 */
const BASIN_NAME_REGEX = /^[a-z0-9][a-z0-9-]{6,46}[a-z0-9]$/;
/**
 * Top-level S2 SDK client.
 *
 * - Authenticates with an access token and exposes account-scoped helpers for basins, streams, access tokens and metrics.
 */
class S2 {
    accessToken;
    client;
    endpoints;
    retryConfig;
    /**
     * Account-scoped basin management operations.
     *
     * - List, create, delete and reconfigure basins.
     */
    basins;
    /** Manage access tokens for the account (list, issue, revoke). */
    accessTokens;
    /** Account, basin and stream level metrics. */
    metrics;
    /**
     * Create a new S2 client.
     *
     * @param options Access token configuration.
     */
    constructor(options) {
        this.accessToken = Redacted.make(options.accessToken);
        // Merge timeout config: top-level options take precedence over retry.* for backwards compatibility
        this.retryConfig = {
            ...options.retry,
            ...(options.requestTimeoutMillis !== undefined && {
                requestTimeoutMillis: options.requestTimeoutMillis,
            }),
            ...(options.connectionTimeoutMillis !== undefined && {
                connectionTimeoutMillis: options.connectionTimeoutMillis,
            }),
        };
        this.endpoints =
            options.endpoints instanceof endpoints_js_1.S2Endpoints
                ? options.endpoints
                : new endpoints_js_1.S2Endpoints(options.endpoints);
        const headers = {};
        if ((0, runtime_js_1.canSetUserAgentHeader)()) {
            headers["user-agent"] = runtime_js_1.DEFAULT_USER_AGENT;
        }
        this.client = (0, index_js_1.createClient)((0, index_js_1.createConfig)({
            baseUrl: this.endpoints.accountBaseUrl(),
            auth: () => Redacted.value(this.accessToken),
            headers: headers,
        }));
        this.client.interceptors.error.use((err, res) => {
            return (0, error_js_1.makeServerError)(res, err);
        });
        this.basins = new basins_js_1.S2Basins(this.client, this.retryConfig);
        this.accessTokens = new accessTokens_js_1.S2AccessTokens(this.client, this.retryConfig);
        this.metrics = new metrics_js_1.S2Metrics(this.client, this.retryConfig);
    }
    /**
     * Create a basin-scoped client bound to a specific basin name.
     *
     * @param name Basin name (8-48 characters, lowercase alphanumeric and hyphens, no leading/trailing hyphens).
     * @throws {S2Error} If the basin name is invalid.
     */
    basin(name) {
        if (!BASIN_NAME_REGEX.test(name)) {
            throw new error_js_1.S2Error({
                message: `Invalid basin name: "${name}". Basin names must be 8-48 characters, ` +
                    `contain only lowercase letters, numbers, and hyphens, and cannot start or end with a hyphen.`,
                origin: "sdk",
            });
        }
        return new basin_js_1.S2Basin(name, {
            accessToken: this.accessToken,
            baseUrl: this.endpoints.basinBaseUrl(name),
            includeBasinHeader: this.endpoints.includeBasinHeader,
            retryConfig: this.retryConfig,
        });
    }
}
exports.S2 = S2;
