// @s2-dev/streamstore/dist/cjs/types.js
"use strict";
/**
 * S2 SDK Types
 *
 * All public SDK types are defined here. Types use camelCase field names
 * for idiomatic JavaScript/TypeScript usage.
 *
 * Generated types (snake_case, matching API wire format) are available
 * from "./generated/types.gen.js" as the `API` namespace.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppendInput = exports.MAX_APPEND_BYTES = exports.MAX_APPEND_RECORDS = exports.AppendRecord = void 0;
const error_js_1 = require("./error.js");
const utils_js_1 = require("./utils.js");
const textEncoder = new TextEncoder();
/**
 * Factory functions for creating AppendRecord instances.
 */
var AppendRecord;
(function (AppendRecord) {
    /**
     * Create a string-format append record with pre-calculated metered size.
     */
    function string(params) {
        // Create record with placeholder, then calculate actual size
        const record = {
            body: params.body,
            headers: params.headers,
            timestamp: params.timestamp,
            meteredBytes: 0,
        };
        // Safe to cast: we're setting a readonly property during construction
        record.meteredBytes =
            (0, utils_js_1.meteredBytes)(record);
        return record;
    }
    AppendRecord.string = string;
    /**
     * Create a bytes-format append record with pre-calculated metered size.
     */
    function bytes(params) {
        // Create record with placeholder, then calculate actual size
        const record = {
            body: params.body,
            headers: params.headers,
            timestamp: params.timestamp,
            meteredBytes: 0,
        };
        // Safe to cast: we're setting a readonly property during construction
        record.meteredBytes =
            (0, utils_js_1.meteredBytes)(record);
        return record;
    }
    AppendRecord.bytes = bytes;
    /**
     * Create a fence command record.
     */
    function fence(fencingToken, timestamp) {
        return string({
            body: fencingToken,
            headers: [["", "fence"]],
            timestamp,
        });
    }
    AppendRecord.fence = fence;
    /**
     * Create a trim command record.
     */
    function trim(seqNum, timestamp) {
        const buffer = new Uint8Array(8);
        const view = new DataView(buffer.buffer);
        view.setBigUint64(0, BigInt(seqNum), false);
        return bytes({
            body: buffer,
            headers: [[textEncoder.encode(""), textEncoder.encode("trim")]],
            timestamp,
        });
    }
    AppendRecord.trim = trim;
})(AppendRecord || (exports.AppendRecord = AppendRecord = {}));
// =============================================================================
// Append Input
// =============================================================================
/** Maximum number of records in a single append batch. */
exports.MAX_APPEND_RECORDS = 1000;
/** Maximum total metered bytes for records in a single append batch (1 MiB). */
exports.MAX_APPEND_BYTES = 1024 * 1024;
/**
 * Factory functions for creating AppendInput instances.
 */
var AppendInput;
(function (AppendInput) {
    /**
     * Create an AppendInput with validation.
     *
     * @throws {S2Error} If validation fails (empty, too many records, or too large)
     */
    function create(records, options) {
        if (records.length === 0) {
            throw new error_js_1.S2Error({
                message: "AppendInput must contain at least one record",
                origin: "sdk",
            });
        }
        if (records.length > exports.MAX_APPEND_RECORDS) {
            throw new error_js_1.S2Error({
                message: `AppendInput cannot contain more than ${exports.MAX_APPEND_RECORDS} records (got ${records.length})`,
                origin: "sdk",
            });
        }
        const totalBytes = records.reduce((sum, r) => sum + r.meteredBytes, 0);
        if (totalBytes > exports.MAX_APPEND_BYTES) {
            throw new error_js_1.S2Error({
                message: `AppendInput exceeds maximum of ${exports.MAX_APPEND_BYTES} bytes (got ${totalBytes} bytes)`,
                origin: "sdk",
            });
        }
        if (options?.fencingToken !== undefined) {
            const tokenBytes = (0, utils_js_1.utf8ByteLength)(options.fencingToken);
            if (tokenBytes > 36) {
                throw new error_js_1.S2Error({
                    message: "fencing token must not exceed 36 bytes in length",
                    origin: "sdk",
                    status: 422,
                });
            }
        }
        return {
            records,
            matchSeqNum: options?.matchSeqNum === undefined
                ? undefined
                : Math.floor(options.matchSeqNum),
            fencingToken: options?.fencingToken,
            meteredBytes: totalBytes,
        };
    }
    AppendInput.create = create;
})(AppendInput || (exports.AppendInput = AppendInput = {}));
